fn main() {
    1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1 = 1;
    //~^ ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
    //~| ERROR invalid left-hand side of assignment
}
