error[E0438]: const `BAR` is not a member of trait `A`
  --> $DIR/trait-impl-can-not-have-untraitful-items.rs:15:5
   |
LL |     const BAR: () = (); //~ ERROR const `BAR` is not a member of trait `A`
   |     ^^^^^^^^^^^^^^^^^^^ not a member of trait `A`

error[E0437]: type `Baz` is not a member of trait `A`
  --> $DIR/trait-impl-can-not-have-untraitful-items.rs:16:5
   |
LL |     type Baz = (); //~ ERROR type `Baz` is not a member of trait `A`
   |     ^^^^^^^^^^^^^^ not a member of trait `A`

error[E0407]: method `foo` is not a member of trait `A`
  --> $DIR/trait-impl-can-not-have-untraitful-items.rs:17:5
   |
LL |     fn foo(&self) { } //~ ERROR method `foo` is not a member of trait `A`
   |     ^^^^^^^^^^^^^^^^^ not a member of trait `A`

error: aborting due to 3 previous errors

Some errors occurred: E0407, E0437, E0438.
For more information about an error, try `rustc --explain E0407`.
