error[E0594]: cannot assign to `*s.pointer` which is behind a `&` reference
  --> $DIR/borrowck-assign-to-andmut-in-aliasable-loc.rs:19:5
   |
LL | fn a(s: &S) {
   |         -- help: consider changing this to be a mutable reference: `&mut S<'_>`
LL |     *s.pointer += 1; //~ ERROR cannot assign
   |     ^^^^^^^^^^^^^^^ `s` is a `&` reference, so the data it refers to cannot be written

error[E0594]: cannot assign to `*s.pointer` which is behind a `&` reference
  --> $DIR/borrowck-assign-to-andmut-in-aliasable-loc.rs:27:5
   |
LL | fn c(s: & &mut S) {
   |         -------- help: consider changing this to be a mutable reference: `&mut &mut S<'_>`
LL |     *s.pointer += 1; //~ ERROR cannot assign
   |     ^^^^^^^^^^^^^^^ `s` is a `&` reference, so the data it refers to cannot be written

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0594`.
