error[E0389]: cannot assign to data in a `&` reference
  --> $DIR/borrowck-assign-to-andmut-in-aliasable-loc.rs:19:5
   |
LL | fn a(s: &S) {
   |         -- use `&mut S` here to make mutable
LL |     *s.pointer += 1; //~ ERROR cannot assign
   |     ^^^^^^^^^^^^^^^ assignment into an immutable reference

error[E0389]: cannot assign to data in a `&` reference
  --> $DIR/borrowck-assign-to-andmut-in-aliasable-loc.rs:27:5
   |
LL | fn c(s: & &mut S) {
   |         -------- use `&mut &mut S` here to make mutable
LL |     *s.pointer += 1; //~ ERROR cannot assign
   |     ^^^^^^^^^^^^^^^ assignment into an immutable reference

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0389`.
