error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:29:17
   |
LL |     let bar1 = &mut foo.bar1;
   |                ------------- first mutable borrow occurs here
LL |     let _bar2 = &mut foo.bar1;  //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^^^ second mutable borrow occurs here
LL |     *bar1;
   |     ----- borrow later used here

error[E0502]: cannot borrow `foo.bar1` as immutable because it is also borrowed as mutable
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:36:17
   |
LL |     let bar1 = &mut foo.bar1;
   |                ------------- mutable borrow occurs here
LL |     let _bar2 = &foo.bar1;  //~ ERROR cannot borrow
   |                 ^^^^^^^^^ immutable borrow occurs here
LL |     *bar1;
   |     ----- borrow later used here

error[E0502]: cannot borrow `foo.bar1` as mutable because it is also borrowed as immutable
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:43:17
   |
LL |     let bar1 = &foo.bar1;
   |                --------- immutable borrow occurs here
LL |     let _bar2 = &mut foo.bar1;  //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^^^ mutable borrow occurs here
LL |     *bar1;
   |     ----- borrow later used here

error[E0502]: cannot borrow `*foo` as immutable because it is also borrowed as mutable
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:72:11
   |
LL |     let bar1 = &mut foo.bar1;
   |                ------------- mutable borrow occurs here
LL |     match *foo {
   |           ^^^^ immutable borrow occurs here
...
LL |     *bar1;
   |     ----- borrow later used here

error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:73:21
   |
LL |     let bar1 = &mut foo.bar1;
   |                ------------- first mutable borrow occurs here
LL |     match *foo {
LL |         Foo { bar1: ref mut _bar1, bar2: _ } => {}
   |                     ^^^^^^^^^^^^^ second mutable borrow occurs here
...
LL |     *bar1;
   |     ----- borrow later used here

error[E0502]: cannot borrow `foo.bar1` as immutable because it is also borrowed as mutable
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:82:17
   |
LL |     let bar1 = &mut foo.bar1.int1;
   |                ------------------ mutable borrow occurs here
LL |     let _foo1 = &foo.bar1; //~ ERROR cannot borrow
   |                 ^^^^^^^^^ immutable borrow occurs here
LL |     let _foo2 = &*foo; //~ ERROR cannot borrow
LL |     *bar1;
   |     ----- borrow later used here

error[E0502]: cannot borrow `*foo` as immutable because it is also borrowed as mutable
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:83:17
   |
LL |     let bar1 = &mut foo.bar1.int1;
   |                ------------------ mutable borrow occurs here
LL |     let _foo1 = &foo.bar1; //~ ERROR cannot borrow
LL |     let _foo2 = &*foo; //~ ERROR cannot borrow
   |                 ^^^^^ immutable borrow occurs here
LL |     *bar1;
   |     ----- borrow later used here

error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:90:17
   |
LL |     let bar1 = &mut foo.bar1.int1;
   |                ------------------ first mutable borrow occurs here
LL |     let _foo1 = &mut foo.bar1; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^^^ second mutable borrow occurs here
LL |     *bar1;
   |     ----- borrow later used here

error[E0499]: cannot borrow `*foo` as mutable more than once at a time
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:97:17
   |
LL |     let bar1 = &mut foo.bar1.int1;
   |                ------------------ first mutable borrow occurs here
LL |     let _foo2 = &mut *foo; //~ ERROR cannot borrow
   |                 ^^^^^^^^^ second mutable borrow occurs here
LL |     *bar1;
   |     ----- borrow later used here

error[E0502]: cannot borrow `foo.bar1` as mutable because it is also borrowed as immutable
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:104:17
   |
LL |     let bar1 = &foo.bar1.int1;
   |                -------------- immutable borrow occurs here
LL |     let _foo1 = &mut foo.bar1; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^^^ mutable borrow occurs here
LL |     *bar1;
   |     ----- borrow later used here

error[E0502]: cannot borrow `*foo` as mutable because it is also borrowed as immutable
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:111:17
   |
LL |     let bar1 = &foo.bar1.int1;
   |                -------------- immutable borrow occurs here
LL |     let _foo2 = &mut *foo; //~ ERROR cannot borrow
   |                 ^^^^^^^^^ mutable borrow occurs here
LL |     *bar1;
   |     ----- borrow later used here

error[E0596]: cannot borrow `foo.bar1` as mutable, as `foo` is not declared as mutable
  --> $DIR/borrowck-borrow-from-owned-ptr.rs:132:16
   |
LL |     let foo = make_foo();
   |         --- help: consider changing this to be mutable: `mut foo`
LL |     let bar1 = &mut foo.bar1; //~ ERROR cannot borrow
   |                ^^^^^^^^^^^^^ cannot borrow as mutable

error: aborting due to 12 previous errors

Some errors occurred: E0499, E0502, E0596.
For more information about an error, try `rustc --explain E0499`.
