error[E0499]: cannot borrow `*arg` as mutable more than once at a time
  --> $DIR/mut-borrow-in-loop.rs:20:25
   |
20 |             (self.func)(arg) //~ ERROR cannot borrow
   |                         ^^^ mutable borrow starts here in previous iteration of loop
21 |         }
22 |     }
   |     - mutable borrow ends here

error[E0499]: cannot borrow `*arg` as mutable more than once at a time
  --> $DIR/mut-borrow-in-loop.rs:26:25
   |
26 |             (self.func)(arg) //~ ERROR cannot borrow
   |                         ^^^ mutable borrow starts here in previous iteration of loop
27 |         }
28 |     }
   |     - mutable borrow ends here

error[E0499]: cannot borrow `*arg` as mutable more than once at a time
  --> $DIR/mut-borrow-in-loop.rs:33:25
   |
33 |             (self.func)(arg) //~ ERROR cannot borrow
   |                         ^^^ mutable borrow starts here in previous iteration of loop
34 |         }
35 |     }
   |     - mutable borrow ends here

error: aborting due to 3 previous errors

