error[E0594]: cannot assign to `x.0`, as `x` is not declared as mutable
  --> $DIR/reassignment_immutable_fields.rs:7:5
   |
LL |     let x: (u32, u32);
   |         - help: consider changing this to be mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
   |     ^^^^^^^ cannot assign

error[E0594]: cannot assign to `x.1`, as `x` is not declared as mutable
  --> $DIR/reassignment_immutable_fields.rs:8:5
   |
LL |     let x: (u32, u32);
   |         - help: consider changing this to be mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
LL |     x.1 = 22; //~ ERROR
   |     ^^^^^^^^ cannot assign

error[E0594]: cannot assign to `x.0`, as `x` is not declared as mutable
  --> $DIR/reassignment_immutable_fields.rs:15:5
   |
LL |     let x: (u32, u32);
   |         - help: consider changing this to be mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
   |     ^^^^^^^ cannot assign

error[E0594]: cannot assign to `x.1`, as `x` is not declared as mutable
  --> $DIR/reassignment_immutable_fields.rs:16:5
   |
LL |     let x: (u32, u32);
   |         - help: consider changing this to be mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
LL |     x.1 = 22; //~ ERROR
   |     ^^^^^^^^ cannot assign

error[E0381]: use of possibly uninitialized variable: `x`
  --> $DIR/reassignment_immutable_fields.rs:17:10
   |
LL |     drop(x); //~ ERROR
   |          ^ use of possibly uninitialized `x`

error: aborting due to 5 previous errors

Some errors occurred: E0381, E0594.
For more information about an error, try `rustc --explain E0381`.
