error[E0594]: cannot assign to `x.a`, as `x` is not declared as mutable
  --> $DIR/reassignment_immutable_fields_overlapping.rs:12:5
   |
LL |     let x: Foo;
   |         - help: consider changing this to be mutable: `mut x`
LL |     x.a = 1;  //~ ERROR
   |     ^^^^^^^ cannot assign

error[E0594]: cannot assign to `x.b`, as `x` is not declared as mutable
  --> $DIR/reassignment_immutable_fields_overlapping.rs:13:5
   |
LL |     let x: Foo;
   |         - help: consider changing this to be mutable: `mut x`
LL |     x.a = 1;  //~ ERROR
LL |     x.b = 22; //~ ERROR
   |     ^^^^^^^^ cannot assign

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0594`.
