error[E0601]: main function not found

error[E0518]: attribute should be applied to function
  --> $DIR/issue-43106-gating-of-inline.rs:21:1
   |
LL |   #[inline = "2100"]
   |   ^^^^^^^^^^^^^^^^^^
LL |   //~^ ERROR attribute should be applied to function
LL | / mod inline {
LL | |     mod inner { #![inline="2100"] }
LL | |     //~^ ERROR attribute should be applied to function
LL | |
...  |
LL | |     //~^ ERROR attribute should be applied to function
LL | | }
   | |_- not a function

error[E0518]: attribute should be applied to function
  --> $DIR/issue-43106-gating-of-inline.rs:24:17
   |
LL |     mod inner { #![inline="2100"] }
   |     ------------^^^^^^^^^^^^^^^^^-- not a function

error[E0518]: attribute should be applied to function
  --> $DIR/issue-43106-gating-of-inline.rs:29:5
   |
LL |     #[inline = "2100"] struct S;
   |     ^^^^^^^^^^^^^^^^^^ --------- not a function

error[E0518]: attribute should be applied to function
  --> $DIR/issue-43106-gating-of-inline.rs:32:5
   |
LL |     #[inline = "2100"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^ ----------- not a function

error[E0518]: attribute should be applied to function
  --> $DIR/issue-43106-gating-of-inline.rs:35:5
   |
LL |     #[inline = "2100"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^ ---------- not a function

error: aborting due to 6 previous errors

You've got a few errors: E0518, E0601
If you want more information on an error, try using "rustc --explain E0518"
