warning: unknown lint: `x5400`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:38:9
   |
LL | #![warn(x5400)] //~ WARN unknown lint: `x5400`
   |         ^^^^^
   |
note: lint level defined here
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:33:28
   |
LL | #![warn(unused_attributes, unknown_lints)]
   |                            ^^^^^^^^^^^^^

warning: unknown lint: `x5300`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:39:10
   |
LL | #![allow(x5300)] //~ WARN unknown lint: `x5300`
   |          ^^^^^

warning: unknown lint: `x5200`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:40:11
   |
LL | #![forbid(x5200)] //~ WARN unknown lint: `x5200`
   |           ^^^^^

warning: unknown lint: `x5100`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:41:9
   |
LL | #![deny(x5100)] //~ WARN unknown lint: `x5100`
   |         ^^^^^

warning: unknown lint: `x5400`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:101:8
   |
LL | #[warn(x5400)]
   |        ^^^^^

warning: unknown lint: `x5400`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:104:25
   |
LL |     mod inner { #![warn(x5400)] }
   |                         ^^^^^

warning: unknown lint: `x5400`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:107:12
   |
LL |     #[warn(x5400)] fn f() { }
   |            ^^^^^

warning: unknown lint: `x5400`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:110:12
   |
LL |     #[warn(x5400)] struct S;
   |            ^^^^^

warning: unknown lint: `x5400`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:113:12
   |
LL |     #[warn(x5400)] type T = S;
   |            ^^^^^

warning: unknown lint: `x5400`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:116:12
   |
LL |     #[warn(x5400)] impl S { }
   |            ^^^^^

warning: unknown lint: `x5300`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:120:9
   |
LL | #[allow(x5300)]
   |         ^^^^^

warning: unknown lint: `x5300`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:123:26
   |
LL |     mod inner { #![allow(x5300)] }
   |                          ^^^^^

warning: unknown lint: `x5300`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:126:13
   |
LL |     #[allow(x5300)] fn f() { }
   |             ^^^^^

warning: unknown lint: `x5300`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:129:13
   |
LL |     #[allow(x5300)] struct S;
   |             ^^^^^

warning: unknown lint: `x5300`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:132:13
   |
LL |     #[allow(x5300)] type T = S;
   |             ^^^^^

warning: unknown lint: `x5300`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:135:13
   |
LL |     #[allow(x5300)] impl S { }
   |             ^^^^^

warning: unknown lint: `x5200`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:139:10
   |
LL | #[forbid(x5200)]
   |          ^^^^^

warning: unknown lint: `x5200`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:142:27
   |
LL |     mod inner { #![forbid(x5200)] }
   |                           ^^^^^

warning: unknown lint: `x5200`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:145:14
   |
LL |     #[forbid(x5200)] fn f() { }
   |              ^^^^^

warning: unknown lint: `x5200`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:148:14
   |
LL |     #[forbid(x5200)] struct S;
   |              ^^^^^

warning: unknown lint: `x5200`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:151:14
   |
LL |     #[forbid(x5200)] type T = S;
   |              ^^^^^

warning: unknown lint: `x5200`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:154:14
   |
LL |     #[forbid(x5200)] impl S { }
   |              ^^^^^

warning: unknown lint: `x5100`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:158:8
   |
LL | #[deny(x5100)]
   |        ^^^^^

warning: unknown lint: `x5100`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:161:25
   |
LL |     mod inner { #![deny(x5100)] }
   |                         ^^^^^

warning: unknown lint: `x5100`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:164:12
   |
LL |     #[deny(x5100)] fn f() { }
   |            ^^^^^

warning: unknown lint: `x5100`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:167:12
   |
LL |     #[deny(x5100)] struct S;
   |            ^^^^^

warning: unknown lint: `x5100`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:170:12
   |
LL |     #[deny(x5100)] type T = S;
   |            ^^^^^

warning: unknown lint: `x5100`
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:173:12
   |
LL |     #[deny(x5100)] impl S { }
   |            ^^^^^

warning: macro_escape is a deprecated synonym for macro_use
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:457:1
   |
LL | #[macro_escape]
   | ^^^^^^^^^^^^^^^

warning: macro_escape is a deprecated synonym for macro_use
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:460:17
   |
LL |     mod inner { #![macro_escape] }
   |                 ^^^^^^^^^^^^^^^^
   |
   = help: consider an outer attribute, #[macro_use] mod ...

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:181:5
   |
LL |     #[macro_use] fn f() { }
   |     ^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:33:9
   |
LL | #![warn(unused_attributes, unknown_lints)]
   |         ^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:184:5
   |
LL |     #[macro_use] struct S;
   |     ^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:187:5
   |
LL |     #[macro_use] type T = S;
   |     ^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:190:5
   |
LL |     #[macro_use] impl S { }
   |     ^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:197:17
   |
LL |     mod inner { #![macro_export] }
   |                 ^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:200:5
   |
LL |     #[macro_export] fn f() { }
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:203:5
   |
LL |     #[macro_export] struct S;
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:206:5
   |
LL |     #[macro_export] type T = S;
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:209:5
   |
LL |     #[macro_export] impl S { }
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:194:1
   |
LL | #[macro_export]
   | ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:216:17
   |
LL |     mod inner { #![plugin_registrar] }
   |                 ^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:221:5
   |
LL |     #[plugin_registrar] struct S;
   |     ^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:224:5
   |
LL |     #[plugin_registrar] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:227:5
   |
LL |     #[plugin_registrar] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:213:1
   |
LL | #[plugin_registrar]
   | ^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:234:17
   |
LL |     mod inner { #![main] }
   |                 ^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:239:5
   |
LL |     #[main] struct S;
   |     ^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:242:5
   |
LL |     #[main] type T = S;
   |     ^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:245:5
   |
LL |     #[main] impl S { }
   |     ^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:231:1
   |
LL | #[main]
   | ^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:252:17
   |
LL |     mod inner { #![start] }
   |                 ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:257:5
   |
LL |     #[start] struct S;
   |     ^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:260:5
   |
LL |     #[start] type T = S;
   |     ^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:263:5
   |
LL |     #[start] impl S { }
   |     ^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:249:1
   |
LL | #[start]
   | ^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:316:5
   |
LL |     #[path = "3800"] fn f() { }
   |     ^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:319:5
   |
LL |     #[path = "3800"]  struct S;
   |     ^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:322:5
   |
LL |     #[path = "3800"] type T = S;
   |     ^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:325:5
   |
LL |     #[path = "3800"] impl S { }
   |     ^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:332:17
   |
LL |     mod inner { #![automatically_derived] }
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:335:5
   |
LL |     #[automatically_derived] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:338:5
   |
LL |     #[automatically_derived] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:341:5
   |
LL |     #[automatically_derived] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:344:5
   |
LL |     #[automatically_derived] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:329:1
   |
LL | #[automatically_derived]
   | ^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:364:17
   |
LL |     mod inner { #![no_link] }
   |                 ^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:367:5
   |
LL |     #[no_link] fn f() { }
   |     ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:370:5
   |
LL |     #[no_link] struct S;
   |     ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:373:5
   |
LL |     #[no_link]type T = S;
   |     ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:376:5
   |
LL |     #[no_link] impl S { }
   |     ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:361:1
   |
LL | #[no_link]
   | ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:383:17
   |
LL |     mod inner { #![should_panic] }
   |                 ^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:386:5
   |
LL |     #[should_panic] fn f() { }
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:389:5
   |
LL |     #[should_panic] struct S;
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:392:5
   |
LL |     #[should_panic] type T = S;
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:395:5
   |
LL |     #[should_panic] impl S { }
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:380:1
   |
LL | #[should_panic]
   | ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:402:17
   |
LL |     mod inner { #![ignore] }
   |                 ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:405:5
   |
LL |     #[ignore] fn f() { }
   |     ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:408:5
   |
LL |     #[ignore] struct S;
   |     ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:411:5
   |
LL |     #[ignore] type T = S;
   |     ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:414:5
   |
LL |     #[ignore] impl S { }
   |     ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:399:1
   |
LL | #[ignore]
   | ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:421:17
   |
LL |     mod inner { #![no_implicit_prelude] }
   |                 ^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:424:5
   |
LL |     #[no_implicit_prelude] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:427:5
   |
LL |     #[no_implicit_prelude] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:430:5
   |
LL |     #[no_implicit_prelude] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:433:5
   |
LL |     #[no_implicit_prelude] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:418:1
   |
LL | #[no_implicit_prelude]
   | ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:440:17
   |
LL |     mod inner { #![reexport_test_harness_main="2900"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:443:5
   |
LL |     #[reexport_test_harness_main = "2900"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:446:5
   |
LL |     #[reexport_test_harness_main = "2900"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:449:5
   |
LL |     #[reexport_test_harness_main = "2900"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:452:5
   |
LL |     #[reexport_test_harness_main = "2900"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:437:1
   |
LL | #[reexport_test_harness_main = "2900"]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:463:5
   |
LL |     #[macro_escape] fn f() { }
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:466:5
   |
LL |     #[macro_escape] struct S;
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:469:5
   |
LL |     #[macro_escape] type T = S;
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:472:5
   |
LL |     #[macro_escape] impl S { }
   |     ^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:480:17
   |
LL |     mod inner { #![no_std] }
   |                 ^^^^^^^^^^

warning: crate-level attribute should be in the root module
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:480:17
   |
LL |     mod inner { #![no_std] }
   |                 ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:484:5
   |
LL |     #[no_std] fn f() { }
   |     ^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:484:5
   |
LL |     #[no_std] fn f() { }
   |     ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:488:5
   |
LL |     #[no_std] struct S;
   |     ^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:488:5
   |
LL |     #[no_std] struct S;
   |     ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:492:5
   |
LL |     #[no_std] type T = S;
   |     ^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:492:5
   |
LL |     #[no_std] type T = S;
   |     ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:496:5
   |
LL |     #[no_std] impl S { }
   |     ^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:496:5
   |
LL |     #[no_std] impl S { }
   |     ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:476:1
   |
LL | #[no_std]
   | ^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:476:1
   |
LL | #[no_std]
   | ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:635:17
   |
LL |     mod inner { #![crate_name="0900"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be in the root module
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:635:17
   |
LL |     mod inner { #![crate_name="0900"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:639:5
   |
LL |     #[crate_name = "0900"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:639:5
   |
LL |     #[crate_name = "0900"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:643:5
   |
LL |     #[crate_name = "0900"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:643:5
   |
LL |     #[crate_name = "0900"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:647:5
   |
LL |     #[crate_name = "0900"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:647:5
   |
LL |     #[crate_name = "0900"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:651:5
   |
LL |     #[crate_name = "0900"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:651:5
   |
LL |     #[crate_name = "0900"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:631:1
   |
LL | #[crate_name = "0900"]
   | ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:631:1
   |
LL | #[crate_name = "0900"]
   | ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:660:17
   |
LL |     mod inner { #![crate_type="0800"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be in the root module
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:660:17
   |
LL |     mod inner { #![crate_type="0800"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:664:5
   |
LL |     #[crate_type = "0800"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:664:5
   |
LL |     #[crate_type = "0800"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:668:5
   |
LL |     #[crate_type = "0800"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:668:5
   |
LL |     #[crate_type = "0800"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:672:5
   |
LL |     #[crate_type = "0800"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:672:5
   |
LL |     #[crate_type = "0800"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:676:5
   |
LL |     #[crate_type = "0800"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:676:5
   |
LL |     #[crate_type = "0800"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:656:1
   |
LL | #[crate_type = "0800"]
   | ^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:656:1
   |
LL | #[crate_type = "0800"]
   | ^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:685:17
   |
LL |     mod inner { #![feature(x0600)] }
   |                 ^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be in the root module
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:685:17
   |
LL |     mod inner { #![feature(x0600)] }
   |                 ^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:689:5
   |
LL |     #[feature(x0600)] fn f() { }
   |     ^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:689:5
   |
LL |     #[feature(x0600)] fn f() { }
   |     ^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:693:5
   |
LL |     #[feature(x0600)] struct S;
   |     ^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:693:5
   |
LL |     #[feature(x0600)] struct S;
   |     ^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:697:5
   |
LL |     #[feature(x0600)] type T = S;
   |     ^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:697:5
   |
LL |     #[feature(x0600)] type T = S;
   |     ^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:701:5
   |
LL |     #[feature(x0600)] impl S { }
   |     ^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:701:5
   |
LL |     #[feature(x0600)] impl S { }
   |     ^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:681:1
   |
LL | #[feature(x0600)]
   | ^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:681:1
   |
LL | #[feature(x0600)]
   | ^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:711:17
   |
LL |     mod inner { #![no_main] }
   |                 ^^^^^^^^^^^

warning: crate-level attribute should be in the root module
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:711:17
   |
LL |     mod inner { #![no_main] }
   |                 ^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:715:5
   |
LL |     #[no_main] fn f() { }
   |     ^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:715:5
   |
LL |     #[no_main] fn f() { }
   |     ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:719:5
   |
LL |     #[no_main] struct S;
   |     ^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:719:5
   |
LL |     #[no_main] struct S;
   |     ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:723:5
   |
LL |     #[no_main] type T = S;
   |     ^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:723:5
   |
LL |     #[no_main] type T = S;
   |     ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:727:5
   |
LL |     #[no_main] impl S { }
   |     ^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:727:5
   |
LL |     #[no_main] impl S { }
   |     ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:707:1
   |
LL | #[no_main]
   | ^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:707:1
   |
LL | #[no_main]
   | ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:749:17
   |
LL |     mod inner { #![recursion_limit="0200"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be in the root module
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:749:17
   |
LL |     mod inner { #![recursion_limit="0200"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:753:5
   |
LL |     #[recursion_limit="0200"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:753:5
   |
LL |     #[recursion_limit="0200"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:757:5
   |
LL |     #[recursion_limit="0200"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:757:5
   |
LL |     #[recursion_limit="0200"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:761:5
   |
LL |     #[recursion_limit="0200"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:761:5
   |
LL |     #[recursion_limit="0200"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:765:5
   |
LL |     #[recursion_limit="0200"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:765:5
   |
LL |     #[recursion_limit="0200"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:745:1
   |
LL | #[recursion_limit="0200"]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:745:1
   |
LL | #[recursion_limit="0200"]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:774:17
   |
LL |     mod inner { #![type_length_limit="0100"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be in the root module
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:774:17
   |
LL |     mod inner { #![type_length_limit="0100"] }
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:778:5
   |
LL |     #[type_length_limit="0100"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:778:5
   |
LL |     #[type_length_limit="0100"] fn f() { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:782:5
   |
LL |     #[type_length_limit="0100"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:782:5
   |
LL |     #[type_length_limit="0100"] struct S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:786:5
   |
LL |     #[type_length_limit="0100"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:786:5
   |
LL |     #[type_length_limit="0100"] type T = S;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:790:5
   |
LL |     #[type_length_limit="0100"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:790:5
   |
LL |     #[type_length_limit="0100"] impl S { }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:770:1
   |
LL | #[type_length_limit="0100"]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: crate-level attribute should be an inner attribute: add an exclamation mark: #![foo]
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:770:1
   |
LL | #[type_length_limit="0100"]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:43:1
   |
LL | #![macro_export] //~ WARN unused attribute
   | ^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:44:1
   |
LL | #![plugin_registrar] //~ WARN unused attribute
   | ^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:47:1
   |
LL | #![main] //~ WARN unused attribute
   | ^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:48:1
   |
LL | #![start] //~ WARN unused attribute
   | ^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:51:1
   |
LL | #![repr()]
   | ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:53:1
   |
LL | #![path = "3800"] //~ WARN unused attribute
   | ^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:54:1
   |
LL | #![automatically_derived] //~ WARN unused attribute
   | ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:56:1
   |
LL | #![no_link] //~ WARN unused attribute
   | ^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:58:1
   |
LL | #![should_panic] //~ WARN unused attribute
   | ^^^^^^^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:59:1
   |
LL | #![ignore] //~ WARN unused attribute
   | ^^^^^^^^^^

warning: unused attribute
  --> $DIR/issue-43106-gating-of-builtin-attrs.rs:65:1
   |
LL | #![proc_macro_derive()] //~ WARN unused attribute
   | ^^^^^^^^^^^^^^^^^^^^^^^

error: invalid windows subsystem `1000`, only `windows` and `console` are allowed

error: aborting due to previous error

