diff --git a/src/llvm-project/llvm/cmake/modules/HandleLLVMOptions.cmake b/src/llvm-project/llvm/cmake/modules/HandleLLVMOptions.cmake
index 31112308..b4321234 100644
--- a/src/llvm-project/llvm/cmake/modules/HandleLLVMOptions.cmake
+++ b/src/llvm-project/llvm/cmake/modules/HandleLLVMOptions.cmake
@@ -243,8 +243,13 @@
     set(LLVM_HAVE_LINK_VERSION_SCRIPT 1)
   endif()
 elseif(CMAKE_SYSTEM_NAME STREQUAL "Generic")
-  set(LLVM_ON_UNIX 0)
-  set(LLVM_HAVE_LINK_VERSION_SCRIPT 0)
+  if(LLVM_DEFAULT_TARGET_TRIPLE MATCHES "thingos")
+    set(LLVM_ON_UNIX 1)
+    set(LLVM_HAVE_LINK_VERSION_SCRIPT 1)
+  else()
+    set(LLVM_ON_UNIX 0)
+    set(LLVM_HAVE_LINK_VERSION_SCRIPT 0)
+  endif()
 else()
   MESSAGE(SEND_ERROR "Unable to determine platform")
 endif()
