error[E0606]: casting `&'static str` as `*const u8` is invalid
  --> $DIR/const-cast-different-types.rs:12:23
   |
LL | static b: *const u8 = a as *const u8; //~ ERROR casting
   |                       ^^^^^^^^^^^^^^

error[E0606]: casting `&&'static str` as `*const u8` is invalid
  --> $DIR/const-cast-different-types.rs:13:23
   |
LL | static c: *const u8 = &a as *const u8; //~ ERROR casting
   |                       ^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0606`.
