From 50ce3e454d5721cb534a9e9bb73c82246b930bab Mon Sep 17 00:00:00 2001
From: bjorn3 <bjorn3@users.noreply.github.com>
Date: Sat, 19 Jan 2019 11:46:43 +0100
Subject: [PATCH] Disable inline assembly in hint::spin_loop

---
 src/libcore/hint.rs | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/libcore/hint.rs b/src/libcore/hint.rs
index ad5a207..04712b8 100644
--- a/src/libcore/hint.rs
+++ b/src/libcore/hint.rs
@@ -62,6 +62,7 @@ pub unsafe fn unreachable_unchecked() -> ! {
 #[inline]
 #[unstable(feature = "renamed_spin_loop", issue = "55002")]
 pub fn spin_loop() {
+    /*
     #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
     unsafe {
         asm!("pause" ::: "memory" : "volatile");
@@ -71,4 +72,5 @@ pub fn spin_loop() {
     unsafe {
         asm!("yield" ::: "memory" : "volatile");
     }
+    */
 }
-- 
2.17.2 (Apple Git-113)

