From da996dae0b95f986de46a916aca00e03257ba4f9 Mon Sep 17 00:00:00 2001
From: bjorn3 <bjorn3@users.noreply.github.com>
Date: Wed, 30 Jan 2019 14:51:57 +0100
Subject: [PATCH] Patch away bswap usage

---
 src/libcore/num/mod.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/libcore/num/mod.rs b/src/libcore/num/mod.rs
index f928d40..6a146f5 100644
--- a/src/libcore/num/mod.rs
+++ b/src/libcore/num/mod.rs
@@ -2303,7 +2303,7 @@ assert_eq!(m, ", $swapped, ");
             #[stable(feature = "rust1", since = "1.0.0")]
             #[inline]
             pub const fn swap_bytes(self) -> Self {
-                intrinsics::bswap(self as $ActualT) as Self
+                0 // bswap is unsupported by cg_clif
             }
         }
 
-- 
2.11.0

