From f845df4056f5ba16b9f5bd703460c4ac40ea03b9 Mon Sep 17 00:00:00 2001
From: Antoni Boucher <bouanto@zoho.com>
Date: Fri, 26 Aug 2022 20:38:58 -0400
Subject: [PATCH] Edit

---
 library/core/tests/lib.rs | 2 --
 1 file changed, 2 deletions(-)

diff --git a/library/core/tests/lib.rs b/library/core/tests/lib.rs
index 59510d3..179bf26 100644
--- a/library/core/tests/lib.rs
+++ b/library/core/tests/lib.rs
@@ -77,7 +77,6 @@
 #![feature(unwrap_infallible)]
 #![feature(result_into_ok_or_err)]
 #![feature(pointer_byte_offsets)]
-#![feature(portable_simd)]
 #![feature(ptr_metadata)]
 #![feature(once_cell)]
 #![feature(option_result_contains)]
@@ -135,7 +134,6 @@ mod pin;
 mod pin_macro;
 mod ptr;
 mod result;
-mod simd;
 mod slice;
 mod str;
 mod str_lossy;
-- 
2.26.2.7.g19db9cfb68.dirty

