From baa9c656ba61633f2e5785fc47bc8f88ed2f738e Mon Sep 17 00:00:00 2001
From: bjorn3 <17426603+bjorn3@users.noreply.github.com>
Date: Wed, 11 Mar 2026 12:34:47 +0000
Subject: [PATCH] Add missing feature gate

---
 coretests/benches/lib.rs | 1 +
 1 file changed, 1 insertion(+)

diff --git a/coretests/benches/lib.rs b/coretests/benches/lib.rs
index 150b9b3..281aa75 100644
--- a/coretests/benches/lib.rs
+++ b/coretests/benches/lib.rs
@@ -8,6 +8,7 @@
 #![feature(iter_array_chunks)]
 #![feature(iter_next_chunk)]
 #![feature(iter_advance_by)]
+#![feature(num_internals)]
 #![feature(uint_gather_scatter_bits)]
 
 extern crate test;
-- 
2.43.0

