error: no method named `closure` found for type `&Obj<[closure@$DIR/issue-33784.rs:35:43: 35:48]>` in the current scope
  --> $DIR/issue-33784.rs:37:7
   |
37 |     p.closure(); //~ ERROR no method named `closure` found
   |       ^^^^^^^ field, not a method
   |
   = help: use `(p.closure)(...)` if you meant to call the function stored in the `closure` field

error: no method named `fn_ptr` found for type `&&Obj<[closure@$DIR/issue-33784.rs:35:43: 35:48]>` in the current scope
  --> $DIR/issue-33784.rs:41:7
   |
41 |     q.fn_ptr(); //~ ERROR no method named `fn_ptr` found
   |       ^^^^^^ field, not a method
   |
   = help: use `(q.fn_ptr)(...)` if you meant to call the function stored in the `fn_ptr` field

error: no method named `c_fn_ptr` found for type `&D` in the current scope
  --> $DIR/issue-33784.rs:46:7
   |
46 |     s.c_fn_ptr(); //~ ERROR no method named `c_fn_ptr` found
   |       ^^^^^^^^ field, not a method
   |
   = help: use `(s.c_fn_ptr)(...)` if you meant to call the function stored in the `c_fn_ptr` field

error: aborting due to previous error(s)

