From a26747487d398af360dc16344b1f9d2b367720c9 Mon Sep 17 00:00:00 2001
From: bjorn3 <bjorn3@users.noreply.github.com>
Date: Thu, 19 Jul 2018 18:26:56 +0200
Subject: [PATCH] Disable stdsimd in libcore

---
 src/libcore/lib.rs | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/libcore/lib.rs b/src/libcore/lib.rs
index bbe6ae8619..bdeb6ce426 100644
--- a/src/libcore/lib.rs
+++ b/src/libcore/lib.rs
@@ -238,6 +238,7 @@ macro_rules! test_v256 { ($item:item) => {}; }
 macro_rules! test_v512 { ($item:item) => {}; }
 #[allow(unused_macros)]
 macro_rules! vector_impl { ($([$f:ident, $($args:tt)*]),*) => { $($f!($($args)*);)* } }
+/*
 #[path = "../stdsimd/coresimd/mod.rs"]
 #[allow(missing_docs, missing_debug_implementations, dead_code, unused_imports)]
 #[unstable(feature = "stdsimd", issue = "48556")]
@@ -250,3 +251,4 @@ pub use coresimd::simd;
 #[stable(feature = "simd_arch", since = "1.27.0")]
 #[cfg(not(stage0))]
 pub use coresimd::arch;
+*/
--
2.11.0

