From 950bfa9eb7a0eb441601cbe0adc1aefcfab8e031 Mon Sep 17 00:00:00 2001
From: bjorn3 <bjorn3@users.noreply.github.com>
Date: Fri, 27 Jul 2018 19:07:01 +0200
Subject: [PATCH] Remove atomics from libcore

---
 src/libcore/lib.rs         |    1 -
 1 files changed, 1 deletions(-)

diff --git a/src/libcore/lib.rs b/src/libcore/lib.rs
index 3aa79087db..71fcff2e8b 100644
--- a/src/libcore/lib.rs
+++ b/src/libcore/lib.rs
@@ -185,7 +185,6 @@ pub mod borrow;
 pub mod any;
 pub mod array;
 pub mod ascii;
-pub mod sync;
 pub mod cell;
 pub mod char;
 pub mod panic;
2.11.0

