From 3d1907db6e349aa7e6cfe3dd490a9d10db669872 Mon Sep 17 00:00:00 2001
From: bjorn3 <bjorn3@users.noreply.github.com>
Date: Thu, 23 Aug 2018 11:15:01 +0200
Subject: [PATCH] Disable some more unsupported stuff in libcore

---
 src/libcore/num/mod.rs | 8 ++++++--
 1 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/libcore/num/mod.rs b/src/libcore/num/mod.rs
index 5d00949..f077f1d 100644
--- a/src/libcore/num/mod.rs
+++ b/src/libcore/num/mod.rs
@@ -2273,7 +2273,9 @@ assert_eq!(m, ", $swapped, ");
             #[rustc_const_unstable(feature = "const_int_ops")]
             #[inline]
             pub const fn swap_bytes(self) -> Self {
-                unsafe { intrinsics::bswap(self as $ActualT) as Self }
+                // TODO support bswap
+                //unsafe { intrinsics::bswap(self as $ActualT) as Self }
+                0
             }
         }

--
2.11.0
