From d873311b7bd16bac383d71d49f93d0448cfcb0b1 Mon Sep 17 00:00:00 2001
From: bjorn3 <bjorn3@users.noreply.github.com>
Date: Sun, 30 Sep 2018 18:32:50 +0200
Subject: [PATCH] Disable compiler_builtins

---
 src/libcompiler_builtins/src/lib.rs | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/libcompiler_builtins/src/lib.rs b/src/libcompiler_builtins/src/lib.rs
index 9f1dd15..b1eed26 100644
--- a/src/libcompiler_builtins/src/lib.rs
+++ b/src/libcompiler_builtins/src/lib.rs
@@ -34,6 +34,7 @@
 // that follow "x86 naming convention" (e.g. addsf3). Those aeabi intrinsics must adhere to the
 // AAPCS calling convention (`extern "aapcs"`) because that's how LLVM will call them.
 
+/*
 #[cfg(test)]
 extern crate core;
 
@@ -68,3 +69,4 @@ pub mod x86;
 pub mod x86_64;
 
 pub mod probestack;
+*/
-- 
2.17.1 (Apple Git-112)

