error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:27:12
   |
LL |         x: 1,
   |            ^
   |            |
   |            expected f32, found integral variable
   |            help: use a float literal: `1.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:30:12
   |
LL |         y: 2,
   |            ^
   |            |
   |            expected f32, found integral variable
   |            help: use a float literal: `2.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:36:12
   |
LL |         x: 3,
   |            ^
   |            |
   |            expected f32, found integral variable
   |            help: use a float literal: `3.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:39:12
   |
LL |         y: 4,
   |            ^
   |            |
   |            expected f32, found integral variable
   |            help: use a float literal: `4.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:45:12
   |
LL |         x: 5,
   |            ^
   |            |
   |            expected f32, found integral variable
   |            help: use a float literal: `5.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:52:12
   |
LL |         x: 7,
   |            ^
   |            |
   |            expected f32, found integral variable
   |            help: use a float literal: `7.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0107]: wrong number of type arguments: expected 0, found 1
  --> $DIR/structure-constructor-type-mismatch.rs:58:24
   |
LL |     let pt3 = PointF::<i32> { //~ ERROR wrong number of type arguments
   |                        ^^^ unexpected type argument

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:59:12
   |
LL |         x: 9,  //~ ERROR mismatched types
   |            ^
   |            |
   |            expected f32, found integral variable
   |            help: use a float literal: `9.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:60:12
   |
LL |         y: 10, //~ ERROR mismatched types
   |            ^^
   |            |
   |            expected f32, found integral variable
   |            help: use a float literal: `10.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0107]: wrong number of type arguments: expected 0, found 1
  --> $DIR/structure-constructor-type-mismatch.rs:64:18
   |
LL |         PointF::<u32> { .. } => {} //~ ERROR wrong number of type arguments
   |                  ^^^ unexpected type argument

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:64:9
   |
LL |         PointF::<u32> { .. } => {} //~ ERROR wrong number of type arguments
   |         ^^^^^^^^^^^^^^^^^^^^ expected integral variable, found f32
   |
   = note: expected type `Point<{integer}>`
              found type `Point<f32>`

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:69:9
   |
LL |         PointF { .. } => {} //~ ERROR mismatched types
   |         ^^^^^^^^^^^^^ expected integral variable, found f32
   |
   = note: expected type `Point<{integer}>`
              found type `Point<f32>`

error[E0308]: mismatched types
  --> $DIR/structure-constructor-type-mismatch.rs:77:9
   |
LL |         PairF::<u32> { .. } => {} //~ ERROR mismatched types
   |         ^^^^^^^^^^^^^^^^^^^ expected integral variable, found f32
   |
   = note: expected type `Pair<{integer}, {integer}>`
              found type `Pair<f32, u32>`

error: aborting due to 13 previous errors

Some errors occurred: E0107, E0308.
For more information about an error, try `rustc --explain E0107`.
