error: program clause dump
  --> $DIR/lower_env3.rs:15:5
   |
LL |     #[rustc_dump_env_program_clauses] //~ ERROR program clause dump
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: forall<Self> { Implemented(Self: Foo) :- FromEnv(Self: Foo). }

error: program clause dump
  --> $DIR/lower_env3.rs:20:5
   |
LL |     #[rustc_dump_env_program_clauses] //~ ERROR program clause dump
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: forall<Self> { FromEnv(Self: std::marker::Sized) :- FromEnv(Self: std::clone::Clone). }
   = note: forall<Self> { Implemented(Self: std::clone::Clone) :- FromEnv(Self: std::clone::Clone). }
   = note: forall<Self> { Implemented(Self: std::marker::Sized) :- FromEnv(Self: std::marker::Sized). }

error: aborting due to 2 previous errors

