#![allow(foo = "")] //~ ERROR E0452
                    //~| ERROR E0452
                    //~| ERROR E0452
                    //~| ERROR E0452
                    //~| ERROR E0452
                    //~| ERROR E0452
fn main() {
}
