error[E0312]: lifetime of reference outlives lifetime of borrowed content...
  --> $DIR/unboxed-closures-infer-argument-types-two-region-pointers.rs:27:15
   |
LL |         x.set(y); //~ ERROR E0312
   |               ^
   |
note: ...the reference is valid for the anonymous lifetime #3 defined on the body at 26:14...
  --> $DIR/unboxed-closures-infer-argument-types-two-region-pointers.rs:26:14
   |
LL |       doit(0, &|x, y| {
   |  ______________^
LL | |         x.set(y); //~ ERROR E0312
LL | |     });
   | |_____^
note: ...but the borrowed content is only valid for the anonymous lifetime #4 defined on the body at 26:14
  --> $DIR/unboxed-closures-infer-argument-types-two-region-pointers.rs:26:14
   |
LL |       doit(0, &|x, y| {
   |  ______________^
LL | |         x.set(y); //~ ERROR E0312
LL | |     });
   | |_____^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0312`.
