error: suffixed literals are not allowed in attributes
  --> $DIR/malformed-interpolated.rs:5:21
   |
LL |         #[my_attr = $expr] //~ ERROR suffixed literals are not allowed in attributes
   |                     ^^^^^
...
LL | check!(0u8); // ERROR, see above
   | ------------ in this macro invocation
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: unexpected token: `-0`
  --> $DIR/malformed-interpolated.rs:5:19
   |
LL |         #[my_attr = $expr] //~ ERROR suffixed literals are not allowed in attributes
   |                   ^
...
LL | check!(-0); // ERROR, see above
   | ----------- in this macro invocation

error: unexpected token: `0 + 0`
  --> $DIR/malformed-interpolated.rs:5:19
   |
LL |         #[my_attr = $expr] //~ ERROR suffixed literals are not allowed in attributes
   |                   ^
...
LL | check!(0 + 0); // ERROR, see above
   | -------------- in this macro invocation

error: aborting due to 3 previous errors

