From eb703e627e7a84f1cd8d0d87f0f69da1f0acf765 Mon Sep 17 00:00:00 2001
From: bjorn3 <bjorn3@users.noreply.github.com>
Date: Fri, 3 Dec 2021 12:16:30 +0100
Subject: [PATCH] Disable long running tests

---
 library/core/tests/slice.rs | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/slice.rs b/slice.rs
index 8402833..84592e0 100644
--- a/slice.rs
+++ b/slice.rs
@@ -1809,6 +1809,7 @@ fn sort_unstable() {
     assert!(v == [0xDEADBEEF]);
 }
 
+/*
 #[test]
 #[cfg(not(target_arch = "wasm32"))]
 #[cfg_attr(miri, ignore)] // Miri is too slow
@@ -1914,6 +1915,7 @@ fn select_nth_unstable() {
     v.select_nth_unstable(0);
     assert!(v == [0xDEADBEEF]);
 }
+*/
 
 #[test]
 #[should_panic(expected = "index 0 greater than length of slice")]
@@ -2462,6 +2462,7 @@ take_tests! {
 #[cfg(not(miri))] // unused in Miri
 const EMPTY_MAX: &'static [()] = &[(); usize::MAX];
 
+/*
 // can't be a constant due to const mutability rules
 #[cfg(not(miri))] // unused in Miri
 macro_rules! empty_max_mut {
@@ -2485,6 +2486,7 @@ take_tests! {
     (take_mut_oob_max_range_to_inclusive, (..=usize::MAX), None, empty_max_mut!()),
     (take_mut_in_bounds_max_range_from, (usize::MAX..), Some(&mut [] as _), empty_max_mut!()),
 }
+*/
 
 #[test]
 fn test_slice_from_ptr_range() {
-- 
2.26.2.7.g19db9cfb68

