From aacbac292e8b470c8be0e284f3005192f94a0481 Mon Sep 17 00:00:00 2001
From: xonx <119700621+xonx4l@users.noreply.github.com>
Date: Tue, 13 Jan 2026 17:20:06 +0000
Subject: [PATCH] Disable f16 math tests for cranelift

---
 coretests/tests/floats/mod.rs | 26 +++++++++++++-------------
 1 file changed, 13 insertions(+), 13 deletions(-)

diff --git a/coretests/tests/floats/mod.rs b/coretests/tests/floats/mod.rs
index 58892b3daa6..dc3da4a3311 100644
--- a/coretests/tests/floats/mod.rs
+++ b/coretests/tests/floats/mod.rs
@@ -1536,7 +1536,7 @@ fn s_nan() -> Float {
     name: powf,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1559,7 +1559,7 @@ fn s_nan() -> Float {
     name: exp,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1580,7 +1580,7 @@ fn s_nan() -> Float {
     name: exp2,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1600,7 +1600,7 @@ fn s_nan() -> Float {
     name: ln,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1622,7 +1622,7 @@ fn s_nan() -> Float {
     name: log,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1647,7 +1647,7 @@ fn s_nan() -> Float {
     name: log2,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1670,7 +1670,7 @@ fn s_nan() -> Float {
     name: log10,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1694,7 +1694,7 @@ fn s_nan() -> Float {
     name: asinh,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1725,7 +1725,7 @@ fn s_nan() -> Float {
     name: acosh,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1752,7 +1752,7 @@ fn s_nan() -> Float {
     name: atanh,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1778,7 +1778,7 @@ fn s_nan() -> Float {
     name: gamma,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1804,7 +1804,7 @@ fn s_nan() -> Float {
     name: ln_gamma,
     attrs: {
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
@@ -1942,7 +1942,7 @@ fn s_nan() -> Float {
     attrs: {
         // FIXME(f16_f128): add math tests when available
         const: #[cfg(false)],
-        f16: #[cfg(any(miri, target_has_reliable_f16_math))],
+        f16: #[cfg(false)], // FIXME(rust-lang/rustc_codegen_cranelift#1622)
         f128: #[cfg(all(not(miri), target_has_reliable_f128_math))],
     },
     test<Float> {
-- 
2.50.1

