warning: unused import: `m::Tr1 as _`
  --> $DIR/basic.rs:36:9
   |
LL |     use m::Tr1 as _; //~ WARN unused import
   |         ^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/basic.rs:14:9
   |
LL | #![warn(unused_imports, unused_extern_crates)]
   |         ^^^^^^^^^^^^^^

warning: unused import: `S as _`
  --> $DIR/basic.rs:37:9
   |
LL |     use S as _; //~ WARN unused import
   |         ^^^^^^

