error[E0658]: const generics are unstable (see issue #44580)
  --> $DIR/feature-gate-const_generics.rs:1:14
   |
LL | fn foo<const X: ()>() {} //~ ERROR const generics are unstable
   |              ^
   |
   = help: add #![feature(const_generics)] to the crate attributes to enable

error[E0658]: const generics are unstable (see issue #44580)
  --> $DIR/feature-gate-const_generics.rs:3:18
   |
LL | struct Foo<const X: usize>([(); X]); //~ ERROR const generics are unstable
   |                  ^
   |
   = help: add #![feature(const_generics)] to the crate attributes to enable

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
