warning: not reporting region error due to nll
  --> $DIR/ex2a-push-one-existing-name-early-bound.rs:17:12
   |
LL |     x.push(y); //~ ERROR explicit lifetime required
   |            ^

error[E0621]: explicit lifetime required in the type of `y`
  --> $DIR/ex2a-push-one-existing-name-early-bound.rs:17:5
   |
LL | fn baz<'a, 'b, T>(x: &mut Vec<&'a T>, y: &T)
   |                                       - consider changing the type of `y` to `&'a T`
...
LL |     x.push(y); //~ ERROR explicit lifetime required
   |     ^^^^^^^^^ lifetime `'a` required

error: aborting due to previous error

For more information about this error, try `rustc --explain E0621`.
