warning: not reporting region error due to nll
  --> $DIR/ex3-both-anon-regions-both-are-structs.rs:15:12
   |
LL |     x.push(y); //~ ERROR lifetime mismatch
   |            ^

error: unsatisfied lifetime constraints
  --> $DIR/ex3-both-anon-regions-both-are-structs.rs:15:5
   |
LL | fn foo(mut x: Vec<Ref>, y: Ref) {
   |        -----            - has type `Ref<'1>`
   |        |
   |        has type `std::vec::Vec<Ref<'2>>`
LL |     x.push(y); //~ ERROR lifetime mismatch
   |     ^^^^^^^^^ argument requires that `'1` must outlive `'2`

error: aborting due to previous error

