warning: not reporting region error due to nll
  --> $DIR/ex3-both-anon-regions-one-is-struct-3.rs:14:11
   |
LL |     y.b = x; //~ ERROR lifetime mismatch
   |           ^

error: unsatisfied lifetime constraints
  --> $DIR/ex3-both-anon-regions-one-is-struct-3.rs:14:5
   |
LL | fn foo(mut y: Ref, x: &u32) {
   |        -----          - let's call the lifetime of this reference `'1`
   |        |
   |        has type `Ref<'_, '2>`
LL |     y.b = x; //~ ERROR lifetime mismatch
   |     ^^^^^^^ requires that `'1` must outlive `'2`

error: aborting due to previous error

