error[E0423]: expected value, found struct `Empty1`
  --> $DIR/empty-struct-braces-expr.rs:25:14
   |
LL |     let e1 = Empty1; //~ ERROR expected value, found struct `Empty1`
   |              ^^^^^^
   |              |
   |              did you mean `XEmpty2`?
   |              did you mean `Empty1 { /* fields */ }`?

error[E0423]: expected function, found struct `Empty1`
  --> $DIR/empty-struct-braces-expr.rs:26:14
   |
LL |     let e1 = Empty1(); //~ ERROR expected function, found struct `Empty1`
   |              ^^^^^^
   |              |
   |              did you mean `XEmpty2`?
   |              did you mean `Empty1 { /* fields */ }`?

error[E0423]: expected value, found struct variant `E::Empty3`
  --> $DIR/empty-struct-braces-expr.rs:27:14
   |
LL |     let e3 = E::Empty3; //~ ERROR expected value, found struct variant `E::Empty3`
   |              ^^^^^^^^^ did you mean `E::Empty3 { /* fields */ }`?

error[E0423]: expected function, found struct variant `E::Empty3`
  --> $DIR/empty-struct-braces-expr.rs:28:14
   |
LL |     let e3 = E::Empty3(); //~ ERROR expected function, found struct variant `E::Empty3`
   |              ^^^^^^^^^ did you mean `E::Empty3 { /* fields */ }`?

error[E0423]: expected value, found struct `XEmpty1`
  --> $DIR/empty-struct-braces-expr.rs:30:15
   |
LL |     let xe1 = XEmpty1; //~ ERROR expected value, found struct `XEmpty1`
   |               ^^^^^^^
   |               |
   |               did you mean `XEmpty2`?
   |               did you mean `XEmpty1 { /* fields */ }`?

error[E0423]: expected function, found struct `XEmpty1`
  --> $DIR/empty-struct-braces-expr.rs:31:15
   |
LL |     let xe1 = XEmpty1(); //~ ERROR expected function, found struct `XEmpty1`
   |               ^^^^^^^
   |               |
   |               did you mean `XEmpty2`?
   |               did you mean `XEmpty1 { /* fields */ }`?

error[E0599]: no variant named `Empty3` found for type `empty_struct::XE` in the current scope
  --> $DIR/empty-struct-braces-expr.rs:32:15
   |
LL |     let xe3 = XE::Empty3; //~ ERROR no variant named `Empty3` found for type
   |               ^^^^^^^^^^ variant not found in `empty_struct::XE`
   |
   = note: did you mean `empty_struct::XE::XEmpty3`?

error[E0599]: no variant named `Empty3` found for type `empty_struct::XE` in the current scope
  --> $DIR/empty-struct-braces-expr.rs:33:15
   |
LL |     let xe3 = XE::Empty3(); //~ ERROR no variant named `Empty3` found for type
   |               ^^^^^^^^^^ variant not found in `empty_struct::XE`
   |
   = note: did you mean `empty_struct::XE::XEmpty3`?

error: aborting due to 8 previous errors

Some errors occurred: E0423, E0599.
For more information about an error, try `rustc --explain E0423`.
