error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
  --> $DIR/repr-packed-contains-align.rs:28:1
   |
LL | struct SC(SA); //~ ERROR: packed type cannot transitively contain a `[repr(align)]` type
   | ^^^^^^^^^^^^^^

error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
  --> $DIR/repr-packed-contains-align.rs:31:1
   |
LL | struct SD(SB); //~ ERROR: packed type cannot transitively contain a `[repr(align)]` type
   | ^^^^^^^^^^^^^^

error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
  --> $DIR/repr-packed-contains-align.rs:34:1
   |
LL | struct SE(UA); //~ ERROR: packed type cannot transitively contain a `[repr(align)]` type
   | ^^^^^^^^^^^^^^

error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
  --> $DIR/repr-packed-contains-align.rs:37:1
   |
LL | struct SF(UB); //~ ERROR: packed type cannot transitively contain a `[repr(align)]` type
   | ^^^^^^^^^^^^^^

error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
  --> $DIR/repr-packed-contains-align.rs:40:1
   |
LL | / union UC { //~ ERROR: packed type cannot transitively contain a `[repr(align)]` type
LL | |     a: UA
LL | | }
   | |_^

error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
  --> $DIR/repr-packed-contains-align.rs:45:1
   |
LL | / union UD { //~ ERROR: packed type cannot transitively contain a `[repr(align)]` type
LL | |     n: UB
LL | | }
   | |_^

error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
  --> $DIR/repr-packed-contains-align.rs:50:1
   |
LL | / union UE { //~ ERROR: packed type cannot transitively contain a `[repr(align)]` type
LL | |     a: SA
LL | | }
   | |_^

error[E0588]: packed type cannot transitively contain a `[repr(align)]` type
  --> $DIR/repr-packed-contains-align.rs:55:1
   |
LL | / union UF { //~ ERROR: packed type cannot transitively contain a `[repr(align)]` type
LL | |     n: SB
LL | | }
   | |_^

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0588`.
