warning[E0170]: pattern binding `A` is named the same as one of the variants of the type `E`
  --> $DIR/issue-14221.rs:23:13
   |
LL |             A => "A",
   |             ^ help: to match on the variant, qualify the path: `E::A`

warning[E0170]: pattern binding `B` is named the same as one of the variants of the type `E`
  --> $DIR/issue-14221.rs:25:13
   |
LL |             B => "B", //~ ERROR: unreachable pattern
   |             ^ help: to match on the variant, qualify the path: `E::B`

error: unreachable pattern
  --> $DIR/issue-14221.rs:25:13
   |
LL |             A => "A",
   |             - matches any value
LL | //~^ WARN pattern binding `A` is named the same as one of the variants of the type `E`
LL |             B => "B", //~ ERROR: unreachable pattern
   |             ^ unreachable pattern
   |
note: lint level defined here
  --> $DIR/issue-14221.rs:11:9
   |
LL | #![deny(unreachable_patterns)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0170`.
