error[E0426]: use of undeclared label `'b`
  --> $DIR/issue-28109.rs:16:9
   |
LL |         'b //~ ERROR use of undeclared label
   |         ^^ undeclared label `'b`

error[E0426]: use of undeclared label `'c`
  --> $DIR/issue-28109.rs:19:9
   |
LL |         'c //~ ERROR use of undeclared label
   |         ^^ undeclared label `'c`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0426`.
