error[E0603]: constant `B` is private
  --> $DIR/issue-17718-const-privacy.rs:15:5
   |
LL | use a::B; //~ ERROR: constant `B` is private
   |     ^^^^

error[E0603]: constant `BAR` is private
  --> $DIR/issue-17718-const-privacy.rs:18:5
   |
LL |     BAR, //~ ERROR: constant `BAR` is private
   |     ^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0603`.
