error: user substs: Canonical { variables: [], value: [u32] }
  --> $DIR/dump-fn-method.rs:36:13
   |
LL |     let x = foo::<u32>; //~ ERROR [u32]
   |             ^^^^^^^^^^

error: user substs: Canonical { variables: [CanonicalVarInfo { kind: Ty(General) }, CanonicalVarInfo { kind: Ty(General) }], value: [?0, u32, ?1] }
  --> $DIR/dump-fn-method.rs:42:13
   |
LL |     let x = <_ as Bazoom<u32>>::method::<_>; //~ ERROR [?0, u32, ?1]
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: user substs: Canonical { variables: [], value: [u8, u16, u32] }
  --> $DIR/dump-fn-method.rs:46:13
   |
LL |     let x = <u8 as Bazoom<u16>>::method::<u32>; //~ ERROR [u8, u16, u32]
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: user substs: Canonical { variables: [CanonicalVarInfo { kind: Ty(General) }, CanonicalVarInfo { kind: Ty(General) }], value: [?0, ?1, u32] }
  --> $DIR/dump-fn-method.rs:54:5
   |
LL |     y.method::<u32>(44, 66); //~ ERROR [?0, ?1, u32]
   |     ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

