error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:8:9
   |
LL |     y = &x; //~ ERROR
   |         ^^ borrowed value does not live long enough
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:22:13
   |
LL |     let y = &x; //~ ERROR
   |             ^^ borrowed value does not live long enough
...
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:46:27
   |
LL |     let y: &'static u32 = &x; //~ ERROR
   |                           ^^ borrowed value does not live long enough
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:51:27
   |
LL |     let _: &'static u32 = &x; //~ ERROR
   |                           ^^ borrowed value does not live long enough
...
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0716]: temporary value dropped while borrowed
  --> $DIR/patterns.rs:53:41
   |
LL |     let _: Vec<&'static String> = vec![&String::new()];
   |                                         ^^^^^^^^^^^^^ - temporary value is freed at the end of this statement
   |                                         |
   |                                         creates a temporary which is freed while still in use
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0716]: temporary value dropped while borrowed
  --> $DIR/patterns.rs:56:52
   |
LL |     let (_, a): (Vec<&'static String>, _) = (vec![&String::new()], 44);
   |                                                    ^^^^^^^^^^^^^      - temporary value is freed at the end of this statement
   |                                                    |
   |                                                    creates a temporary which is freed while still in use
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0716]: temporary value dropped while borrowed
  --> $DIR/patterns.rs:59:53
   |
LL |     let (_a, b): (Vec<&'static String>, _) = (vec![&String::new()], 44);
   |                                                     ^^^^^^^^^^^^^      - temporary value is freed at the end of this statement
   |                                                     |
   |                                                     creates a temporary which is freed while still in use
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:65:40
   |
LL |     let (_, _): (&'static u32, u32) = (&x, 44); //~ ERROR
   |                                        ^^ borrowed value does not live long enough
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:70:40
   |
LL |     let (y, _): (&'static u32, u32) = (&x, 44); //~ ERROR
   |                                        ^^ borrowed value does not live long enough
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:75:69
   |
LL |     let Single { value: y }: Single<&'static u32> = Single { value: &x }; //~ ERROR
   |                                                                     ^^ borrowed value does not live long enough
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:80:69
   |
LL |     let Single { value: _ }: Single<&'static u32> = Single { value: &x }; //~ ERROR
   |                                                                     ^^ borrowed value does not live long enough
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: `x` does not live long enough
  --> $DIR/patterns.rs:88:17
   |
LL |         value1: &x, //~ ERROR
   |                 ^^ borrowed value does not live long enough
...
LL | }
   | - `x` dropped here while still borrowed
   |
   = note: borrowed value must be valid for the static lifetime...

error: unsatisfied lifetime constraints
  --> $DIR/patterns.rs:101:5
   |
LL | fn static_to_a_to_static_through_variable<'a>(x: &'a u32) -> &'static u32 {
   |                                           -- lifetime `'a` defined here
...
LL |     y //~ ERROR
   |     ^ returning this value requires that `'a` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/patterns.rs:117:18
   |
LL | fn a_to_static_then_static<'a>(x: &'a u32) -> &'static u32 {
   |                            -- lifetime `'a` defined here
LL |     let (y, _z): (&'static u32, u32) = (x, 44); //~ ERROR
   |                  ^^^^^^^^^^^^^^^^^^^ type annotation requires that `'a` must outlive `'static`

error: aborting due to 14 previous errors

Some errors occurred: E0597, E0716.
For more information about an error, try `rustc --explain E0597`.
