error[E0606]: casting `E` as `f32` is invalid
  --> $DIR/enum-to-float-cast.rs:23:18
   |
LL | static C0: f32 = E::L0 as f32; //~ ERROR casting
   |                  ^^^^^^^^^^^^
   |
   = help: cast through an integer first

error[E0606]: casting `F` as `f32` is invalid
  --> $DIR/enum-to-float-cast.rs:24:18
   |
LL | static C1: f32 = F::H1 as f32; //~ ERROR casting
   |                  ^^^^^^^^^^^^
   |
   = help: cast through an integer first

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0606`.
