error[E0507]: cannot move out of borrowed content
  --> $DIR/access-mode-in-closures.rs:18:15
   |
LL |         match *s { S(v) => v } //~ ERROR cannot move out
   |               ^^     - data moved here
   |               |
   |               cannot move out of borrowed content
   |               help: consider removing the `*`: `s`
   |
note: move occurs because `v` has type `std::vec::Vec<isize>`, which does not implement the `Copy` trait
  --> $DIR/access-mode-in-closures.rs:18:22
   |
LL |         match *s { S(v) => v } //~ ERROR cannot move out
   |                      ^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0507`.
