error[E0517]: attribute should be applied to struct, enum or union
  --> $DIR/attr-usage-repr.rs:4:8
   |
LL | #[repr(C)] //~ ERROR: attribute should be applied to struct, enum or union
   |        ^
LL | fn f() {}
   | --------- not a struct, enum or union

error[E0517]: attribute should be applied to enum
  --> $DIR/attr-usage-repr.rs:16:8
   |
LL | #[repr(i8)] //~ ERROR: attribute should be applied to enum
   |        ^^
LL | struct SInt(f64, f64);
   | ---------------------- not an enum

error[E0517]: attribute should be applied to struct or union
  --> $DIR/attr-usage-repr.rs:25:8
   |
LL | #[repr(packed)] //~ ERROR: attribute should be applied to struct
   |        ^^^^^^
LL | enum EPacked { A, B }
   | --------------------- not a struct or union

error[E0517]: attribute should be applied to struct
  --> $DIR/attr-usage-repr.rs:28:8
   |
LL | #[repr(simd)] //~ ERROR: attribute should be applied to struct
   |        ^^^^
LL | enum ESimd { A, B }
   | ------------------- not a struct

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0517`.
