error[E0599]: no method named `test_mut` found for type `std::vec::Vec<{integer}>` in the current scope
  --> $DIR/auto-ref-slice-plus-ref.rs:18:7
   |
LL |     a.test_mut(); //~ ERROR no method named `test_mut` found
   |       ^^^^^^^^
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `test_mut`, perhaps you need to implement it:
           candidate #1: `MyIter`
   = help: did you mean `get_mut`?

error[E0599]: no method named `test` found for type `std::vec::Vec<{integer}>` in the current scope
  --> $DIR/auto-ref-slice-plus-ref.rs:19:7
   |
LL |     a.test(); //~ ERROR no method named `test` found
   |       ^^^^
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `test`, perhaps you need to implement it:
           candidate #1: `MyIter`

error[E0599]: no method named `test` found for type `[{integer}; 1]` in the current scope
  --> $DIR/auto-ref-slice-plus-ref.rs:21:11
   |
LL |     ([1]).test(); //~ ERROR no method named `test` found
   |           ^^^^
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `test`, perhaps you need to implement it:
           candidate #1: `MyIter`

error[E0599]: no method named `test` found for type `&[{integer}; 1]` in the current scope
  --> $DIR/auto-ref-slice-plus-ref.rs:22:12
   |
LL |     (&[1]).test(); //~ ERROR no method named `test` found
   |            ^^^^
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `test`, perhaps you need to implement it:
           candidate #1: `MyIter`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0599`.
