error[E0620]: cast to unsized type: `&{integer}` as `std::marker::Send`
  --> $DIR/cast-to-unsized-trait-object-suggestion.rs:12:5
   |
12 |     &1 as Send; //~ ERROR cast to unsized
   |     ^^^^^^----
   |           |
   |           help: try casting to a reference instead: `&Send`

error[E0620]: cast to unsized type: `std::boxed::Box<{integer}>` as `std::marker::Send`
  --> $DIR/cast-to-unsized-trait-object-suggestion.rs:13:5
   |
13 |     Box::new(1) as Send; //~ ERROR cast to unsized
   |     ^^^^^^^^^^^^^^^----
   |                    |
   |                    help: try casting to a `Box` instead: `Box<Send>`

error: aborting due to 2 previous errors

