error[E0308]: mismatched types
  --> $DIR/coercion-missing-tail-expected-type.rs:13:28
   |
13 |   fn plus_one(x: i32) -> i32 { //~ ERROR mismatched types
   |  ____________________________^
14 | |     x + 1;
   | |          - help: consider removing this semicolon
15 | | }
   | |_^ expected i32, found ()
   |
   = note: expected type `i32`
              found type `()`

error[E0308]: mismatched types
  --> $DIR/coercion-missing-tail-expected-type.rs:17:29
   |
17 |   fn foo() -> Result<u8, u64> { //~ ERROR mismatched types
   |  _____________________________^
18 | |     Ok(1);
   | |          - help: consider removing this semicolon
19 | | }
   | |_^ expected enum `std::result::Result`, found ()
   |
   = note: expected type `std::result::Result<u8, u64>`
              found type `()`

error: aborting due to 2 previous errors

