error[E0658]: `cfg(target_feature)` is experimental and subject to change (see issue #29717)
  --> $DIR/feature-gate-cfg-target-feature.rs:12:12
   |
12 | #[cfg_attr(target_feature = "x", x)] //~ ERROR `cfg(target_feature)` is experimental
   |            ^^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(cfg_target_feature)] to the crate attributes to enable

error[E0658]: `cfg(target_feature)` is experimental and subject to change (see issue #29717)
  --> $DIR/feature-gate-cfg-target-feature.rs:11:7
   |
11 | #[cfg(target_feature = "x")] //~ ERROR `cfg(target_feature)` is experimental
   |       ^^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(cfg_target_feature)] to the crate attributes to enable

error[E0658]: `cfg(target_feature)` is experimental and subject to change (see issue #29717)
  --> $DIR/feature-gate-cfg-target-feature.rs:15:19
   |
15 | #[cfg(not(any(all(target_feature = "x"))))] //~ ERROR `cfg(target_feature)` is experimental
   |                   ^^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(cfg_target_feature)] to the crate attributes to enable

error[E0658]: `cfg(target_feature)` is experimental and subject to change (see issue #29717)
  --> $DIR/feature-gate-cfg-target-feature.rs:19:10
   |
19 |     cfg!(target_feature = "x");
   |          ^^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(cfg_target_feature)] to the crate attributes to enable

error: aborting due to 4 previous errors

