error[E0658]: multiple-element slice matches anywhere but at the end of a slice (e.g. `[0, ..xs, 0]`) are experimental (see issue #23121)
  --> $DIR/feature-gate-advanced-slice-features.rs:18:9
   |
18 |         [ xs.., 4, 5 ] => {}    //~ ERROR multiple-element slice matches
   |         ^^^^^^^^^^^^^^
   |
   = help: add #![feature(advanced_slice_patterns)] to the crate attributes to enable

error[E0658]: multiple-element slice matches anywhere but at the end of a slice (e.g. `[0, ..xs, 0]`) are experimental (see issue #23121)
  --> $DIR/feature-gate-advanced-slice-features.rs:19:9
   |
19 |         [ 1, xs.., 5 ] => {}    //~ ERROR multiple-element slice matches
   |         ^^^^^^^^^^^^^^
   |
   = help: add #![feature(advanced_slice_patterns)] to the crate attributes to enable

error: aborting due to 2 previous errors

