error[E0432]: unresolved import `zed::baz`
  --> $DIR/import.rs:12:5
   |
LL | use zed::baz; //~ ERROR unresolved import `zed::baz` [E0432]
   |     ^^^^^^^^ no `baz` in `zed`. Did you mean to use `bar`?

error[E0432]: unresolved import `foo`
  --> $DIR/import.rs:18:9
   |
LL |     use foo; //~ ERROR unresolved import `foo` [E0432]
   |         ^^^ no `foo` in the root

error[E0603]: unresolved item `foo` is private
  --> $DIR/import.rs:23:5
   |
LL |     zed::foo(); //~ ERROR `foo` is private
   |     ^^^^^^^^

error: aborting due to 3 previous errors

Some errors occurred: E0432, E0603.
For more information about an error, try `rustc --explain E0432`.
