error[E0106]: missing lifetime specifier
  --> $DIR/issue-19707.rs:13:28
   |
13 | type foo = fn(&u8, &u8) -> &u8; //~ ERROR missing lifetime specifier
   |                            ^ expected lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from argument 1 or argument 2

error[E0106]: missing lifetime specifier
  --> $DIR/issue-19707.rs:15:27
   |
15 | fn bar<F: Fn(&u8, &u8) -> &u8>(f: &F) {} //~ ERROR missing lifetime specifier
   |                           ^ expected lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from argument 1 or argument 2

error: aborting due to 2 previous errors

