error[E0616]: field `priv_field` of struct `my_mod::MyStruct` is private
  --> $DIR/issue-3763.rs:25:19
   |
LL |     let _woohoo = (&my_struct).priv_field;
   |                   ^^^^^^^^^^^^^^^^^^^^^^^

error[E0616]: field `priv_field` of struct `my_mod::MyStruct` is private
  --> $DIR/issue-3763.rs:28:19
   |
LL |     let _woohoo = (Box::new(my_struct)).priv_field;
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0624]: method `happyfun` is private
  --> $DIR/issue-3763.rs:31:18
   |
LL |     (&my_struct).happyfun();               //~ ERROR method `happyfun` is private
   |                  ^^^^^^^^

error[E0624]: method `happyfun` is private
  --> $DIR/issue-3763.rs:33:27
   |
LL |     (Box::new(my_struct)).happyfun();          //~ ERROR method `happyfun` is private
   |                           ^^^^^^^^

error[E0616]: field `priv_field` of struct `my_mod::MyStruct` is private
  --> $DIR/issue-3763.rs:34:16
   |
LL |     let nope = my_struct.priv_field;
   |                ^^^^^^^^^^^^^^^^^^^^

error: aborting due to 5 previous errors

Some errors occurred: E0616, E0624.
For more information about an error, try `rustc --explain E0616`.
