error[E0532]: expected tuple struct/variant, found unit variant `E::A`
  --> $DIR/issue-pr29383.rs:19:14
   |
LL |         Some(E::A(..)) => {} //~ ERROR expected tuple struct/variant, found unit variant `E::A`
   |              ^^^^ not a tuple struct/variant

error[E0532]: expected tuple struct/variant, found unit variant `E::B`
  --> $DIR/issue-pr29383.rs:20:14
   |
LL |         Some(E::B(..)) => {} //~ ERROR expected tuple struct/variant, found unit variant `E::B`
   |              ^^^^ not a tuple struct/variant

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0532`.
