error: expected identifier, found keyword `true`
  --> $DIR/issue-44406.rs:18:10
   |
18 |     foo!(true); //~ ERROR expected type, found keyword
   |          ^^^^ expected identifier, found keyword

error: expected type, found keyword `true`
  --> $DIR/issue-44406.rs:18:10
   |
13 |         bar(baz: $rest)
   |                - help: did you mean to use `;` here?
...
18 |     foo!(true); //~ ERROR expected type, found keyword
   |          ^^^^ expecting a type here because of type ascription

