warning: not reporting region error due to nll
  --> $DIR/issue-52213.rs:12:11
   |
LL |     match (&t,) { //~ ERROR cannot infer an appropriate lifetime
   |           ^^^^^

error: unsatisfied lifetime constraints
  --> $DIR/issue-52213.rs:12:11
   |
LL |     match (&t,) { //~ ERROR cannot infer an appropriate lifetime
   |           ^^^^^ free region requires that `'a` must outlive `'b`

error: aborting due to previous error

