error[E0641]: cannot cast to a pointer of an unknown kind
  --> $DIR/issue-45730.rs:13:23
   |
LL |     let x: *const _ = 0 as _; //~ ERROR cannot cast
   |                       ^^^^^-
   |                            |
   |                            help: consider giving more type information
   |
   = note: The type information given here is insufficient to check whether the pointer cast is valid

error[E0641]: cannot cast to a pointer of an unknown kind
  --> $DIR/issue-45730.rs:15:23
   |
LL |     let x: *const _ = 0 as *const _; //~ ERROR cannot cast
   |                       ^^^^^--------
   |                            |
   |                            help: consider giving more type information
   |
   = note: The type information given here is insufficient to check whether the pointer cast is valid

error[E0641]: cannot cast to a pointer of an unknown kind
  --> $DIR/issue-45730.rs:18:13
   |
LL |     let x = 0 as *const i32 as *const _ as *mut _; //~ ERROR cannot cast
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^------
   |                                            |
   |                                            help: consider giving more type information
   |
   = note: The type information given here is insufficient to check whether the pointer cast is valid

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0641`.
