warning: label name `'fl` shadows a label name that is already in scope
  --> $DIR/loops-reject-duplicate-labels-2.rs:23:7
   |
LL |     { 'fl: for _ in 0..10 { break; } }
   |       --- first declared here
LL |     { 'fl: loop { break; } }             //~ WARN label name `'fl` shadows a label name that is already in scope
   |       ^^^ lifetime 'fl already in scope

warning: label name `'lf` shadows a label name that is already in scope
  --> $DIR/loops-reject-duplicate-labels-2.rs:25:7
   |
LL |     { 'lf: loop { break; } }
   |       --- first declared here
LL |     { 'lf: for _ in 0..10 { break; } }   //~ WARN label name `'lf` shadows a label name that is already in scope
   |       ^^^ lifetime 'lf already in scope

warning: label name `'wl` shadows a label name that is already in scope
  --> $DIR/loops-reject-duplicate-labels-2.rs:27:7
   |
LL |     { 'wl: while 2 > 1 { break; } }
   |       --- first declared here
LL |     { 'wl: loop { break; } }             //~ WARN label name `'wl` shadows a label name that is already in scope
   |       ^^^ lifetime 'wl already in scope

warning: label name `'lw` shadows a label name that is already in scope
  --> $DIR/loops-reject-duplicate-labels-2.rs:29:7
   |
LL |     { 'lw: loop { break; } }
   |       --- first declared here
LL |     { 'lw: while 2 > 1 { break; } }      //~ WARN label name `'lw` shadows a label name that is already in scope
   |       ^^^ lifetime 'lw already in scope

warning: label name `'fw` shadows a label name that is already in scope
  --> $DIR/loops-reject-duplicate-labels-2.rs:31:7
   |
LL |     { 'fw: for _ in 0..10 { break; } }
   |       --- first declared here
LL |     { 'fw: while 2 > 1 { break; } }      //~ WARN label name `'fw` shadows a label name that is already in scope
   |       ^^^ lifetime 'fw already in scope

warning: label name `'wf` shadows a label name that is already in scope
  --> $DIR/loops-reject-duplicate-labels-2.rs:33:7
   |
LL |     { 'wf: while 2 > 1 { break; } }
   |       --- first declared here
LL |     { 'wf: for _ in 0..10 { break; } }   //~ WARN label name `'wf` shadows a label name that is already in scope
   |       ^^^ lifetime 'wf already in scope

warning: label name `'tl` shadows a label name that is already in scope
  --> $DIR/loops-reject-duplicate-labels-2.rs:35:7
   |
LL |     { 'tl: while let Some(_) = None::<i32> { break; } }
   |       --- first declared here
LL |     { 'tl: loop { break; } }             //~ WARN label name `'tl` shadows a label name that is already in scope
   |       ^^^ lifetime 'tl already in scope

warning: label name `'lt` shadows a label name that is already in scope
  --> $DIR/loops-reject-duplicate-labels-2.rs:37:7
   |
LL |     { 'lt: loop { break; } }
   |       --- first declared here
LL |     { 'lt: while let Some(_) = None::<i32> { break; } }
   |       ^^^ lifetime 'lt already in scope

error: compilation successful
  --> $DIR/loops-reject-duplicate-labels-2.rs:42:1
   |
LL | / pub fn main() { //~ ERROR compilation successful
LL | |     foo();
LL | | }
   | |_^

