error[E0130]: patterns aren't allowed in foreign function declarations
  --> $DIR/no-patterns-in-args.rs:12:11
   |
12 |     fn f1(mut arg: u8); //~ ERROR patterns aren't allowed in foreign function declarations
   |           ^^^^^^^ pattern not allowed in foreign function

error[E0130]: patterns aren't allowed in foreign function declarations
  --> $DIR/no-patterns-in-args.rs:13:11
   |
13 |     fn f2(&arg: u8); //~ ERROR patterns aren't allowed in foreign function declarations
   |           ^^^^ pattern not allowed in foreign function

error[E0130]: patterns aren't allowed in foreign function declarations
  --> $DIR/no-patterns-in-args.rs:14:11
   |
14 |     fn f3(arg @ _: u8); //~ ERROR patterns aren't allowed in foreign function declarations
   |           ^^^^^^^ pattern not allowed in foreign function

error[E0561]: patterns aren't allowed in function pointer types
  --> $DIR/no-patterns-in-args.rs:20:14
   |
20 | type A1 = fn(mut arg: u8); //~ ERROR patterns aren't allowed in function pointer types
   |              ^^^^^^^

error[E0561]: patterns aren't allowed in function pointer types
  --> $DIR/no-patterns-in-args.rs:21:14
   |
21 | type A2 = fn(&arg: u8); //~ ERROR patterns aren't allowed in function pointer types
   |              ^^^^

error: aborting due to 5 previous errors

