error[E0259]: the name `core` is defined multiple times
  --> $DIR/no-std-inject.rs:14:1
   |
LL | extern crate core; //~ ERROR: the name `core` is defined multiple times
   | ^^^^^^^^^^^^^^^^^^ `core` reimported here
   |
   = note: `core` must be defined only once in the type namespace of this module
help: You can use `as` to change the binding name of the import
   |
LL | extern crate core as other_core; //~ ERROR: the name `core` is defined multiple times
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0259`.
