error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:13:10
   |
LL | #[path = 1usize] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:14:10
   |
LL | #[path = 1u8] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:15:10
   |
LL | #[path = 1u16] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:16:10
   |
LL | #[path = 1u32] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:17:10
   |
LL | #[path = 1u64] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:18:10
   |
LL | #[path = 1isize] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:19:10
   |
LL | #[path = 1i8] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:20:10
   |
LL | #[path = 1i16] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:21:10
   |
LL | #[path = 1i32] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:22:10
   |
LL | #[path = 1i64] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:23:10
   |
LL | #[path = 1.0f32] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: suffixed literals are not allowed in attributes
  --> $DIR/suffixed-literal-meta.rs:24:10
   |
LL | #[path = 1.0f64] //~ ERROR: suffixed literals are not allowed in attributes
   |          ^^^^^^
   |
   = help: instead of using a suffixed literal (1u8, 1.0f32, etc.), use an unsuffixed version (1, 1.0, etc.).

error: aborting due to 12 previous errors

