error[E0632]: cannot provide explicit type parameters when `impl Trait` is used in argument position.
  --> $DIR/synthetic-param.rs:30:5
   |
LL |     func::<u8>(42); //~ ERROR cannot provide explicit type parameters
   |     ^^^^^^^^^^

error[E0632]: cannot provide explicit type parameters when `impl Trait` is used in argument position.
  --> $DIR/synthetic-param.rs:33:5
   |
LL |     Foo::func::<u8>(42); //~ ERROR cannot provide explicit type parameters
   |     ^^^^^^^^^^^^^^^

error[E0632]: cannot provide explicit type parameters when `impl Trait` is used in argument position.
  --> $DIR/synthetic-param.rs:36:5
   |
LL |     Bar::<i8>::func::<u8>(42); //~ ERROR cannot provide explicit type parameters
   |     ^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0632`.
