error[E0277]: the trait bound `W: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:33:8
   |
33 |     VA(W), //~ ERROR `W: std::marker::Sized` is not satisfied
   |        ^^ `W` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `W`
   = help: consider adding a `where W: std::marker::Sized` bound
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `X: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:34:8
   |
34 |     VB{x: X}, //~ ERROR `X: std::marker::Sized` is not satisfied
   |        ^^^^ `X` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `X`
   = help: consider adding a `where X: std::marker::Sized` bound
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `Y: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:35:15
   |
35 |     VC(isize, Y), //~ ERROR `Y: std::marker::Sized` is not satisfied
   |               ^^ `Y` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `Y`
   = help: consider adding a `where Y: std::marker::Sized` bound
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `Z: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:36:18
   |
36 |     VD{u: isize, x: Z}, //~ ERROR `Z: std::marker::Sized` is not satisfied
   |                  ^^^^ `Z` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `Z`
   = help: consider adding a `where Z: std::marker::Sized` bound
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `[u8]: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:39:8
   |
39 |     VE([u8]), //~ ERROR `[u8]: std::marker::Sized` is not satisfied
   |        ^^^^^ `[u8]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `str: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:40:8
   |
40 |     VF{x: str}, //~ ERROR `str: std::marker::Sized` is not satisfied
   |        ^^^^^^ `str` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `str`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `[f32]: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:41:15
   |
41 |     VG(isize, [f32]), //~ ERROR `[f32]: std::marker::Sized` is not satisfied
   |               ^^^^^^ `[f32]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[f32]`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `[u32]: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:42:18
   |
42 |     VH{u: isize, x: [u32]}, //~ ERROR `[u32]: std::marker::Sized` is not satisfied
   |                  ^^^^^^^^ `[u32]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u32]`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `Foo + 'static: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:51:8
   |
51 |     VM(Foo),  //~ ERROR `Foo + 'static: std::marker::Sized` is not satisfied
   |        ^^^^ `Foo + 'static` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `Foo + 'static`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `Bar + 'static: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:52:8
   |
52 |     VN{x: Bar}, //~ ERROR `Bar + 'static: std::marker::Sized` is not satisfied
   |        ^^^^^^ `Bar + 'static` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `Bar + 'static`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `FooBar + 'static: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:53:15
   |
53 |     VO(isize, FooBar), //~ ERROR `FooBar + 'static: std::marker::Sized` is not satisfied
   |               ^^^^^^^ `FooBar + 'static` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `FooBar + 'static`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `BarFoo + 'static: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:54:18
   |
54 |     VP{u: isize, x: BarFoo}, //~ ERROR `BarFoo + 'static: std::marker::Sized` is not satisfied
   |                  ^^^^^^^^^ `BarFoo + 'static` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `BarFoo + 'static`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `[i8]: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:57:8
   |
57 |     VQ(<&'static [i8] as Deref>::Target), //~ ERROR `[i8]: std::marker::Sized` is not satisfied
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `[i8]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[i8]`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `[char]: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:58:8
   |
58 |     VR{x: <&'static [char] as Deref>::Target},
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `[char]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[char]`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `[f64]: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:60:15
   |
60 |     VS(isize, <&'static [f64] as Deref>::Target),
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `[f64]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[f64]`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `[i32]: std::marker::Sized` is not satisfied
  --> $DIR/unsized-enum2.rs:62:18
   |
62 |     VT{u: isize, x: <&'static [i32] as Deref>::Target},
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `[i32]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[i32]`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `PathHelper1 + 'static: std::marker::Sized` is not satisfied in `Path1`
  --> $DIR/unsized-enum2.rs:45:8
   |
45 |     VI(Path1), //~ ERROR `PathHelper1 + 'static: std::marker::Sized` is not satisfied
   |        ^^^^^^ `PathHelper1 + 'static` does not have a constant size known at compile-time
   |
   = help: within `Path1`, the trait `std::marker::Sized` is not implemented for `PathHelper1 + 'static`
   = note: required because it appears within the type `Path1`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `PathHelper2 + 'static: std::marker::Sized` is not satisfied in `Path2`
  --> $DIR/unsized-enum2.rs:46:8
   |
46 |     VJ{x: Path2}, //~ ERROR `PathHelper2 + 'static: std::marker::Sized` is not satisfied
   |        ^^^^^^^^ `PathHelper2 + 'static` does not have a constant size known at compile-time
   |
   = help: within `Path2`, the trait `std::marker::Sized` is not implemented for `PathHelper2 + 'static`
   = note: required because it appears within the type `Path2`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `PathHelper3 + 'static: std::marker::Sized` is not satisfied in `Path3`
  --> $DIR/unsized-enum2.rs:47:15
   |
47 |     VK(isize, Path3), //~ ERROR `PathHelper3 + 'static: std::marker::Sized` is not satisfied
   |               ^^^^^^ `PathHelper3 + 'static` does not have a constant size known at compile-time
   |
   = help: within `Path3`, the trait `std::marker::Sized` is not implemented for `PathHelper3 + 'static`
   = note: required because it appears within the type `Path3`
   = note: no field of an enum variant may have a dynamically sized type

error[E0277]: the trait bound `PathHelper4 + 'static: std::marker::Sized` is not satisfied in `Path4`
  --> $DIR/unsized-enum2.rs:48:18
   |
48 |     VL{u: isize, x: Path4}, //~ ERROR `PathHelper4 + 'static: std::marker::Sized` is not satisfied
   |                  ^^^^^^^^ `PathHelper4 + 'static` does not have a constant size known at compile-time
   |
   = help: within `Path4`, the trait `std::marker::Sized` is not implemented for `PathHelper4 + 'static`
   = note: required because it appears within the type `Path4`
   = note: no field of an enum variant may have a dynamically sized type

error: aborting due to 20 previous errors

