error[E0165]: irrefutable while-let pattern
  --> $DIR/while-let.rs:23:10
   |
LL |     foo!(a, 1, { //~ ERROR irrefutable while-let
   |          ^ irrefutable pattern

error[E0165]: irrefutable while-let pattern
  --> $DIR/while-let.rs:26:10
   |
LL |     bar!(a, 1, { //~ ERROR irrefutable while-let
   |          ^ irrefutable pattern

error[E0165]: irrefutable while-let pattern
  --> $DIR/while-let.rs:32:15
   |
LL |     while let a = 1 { //~ ERROR irrefutable while-let
   |               ^ irrefutable pattern

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0165`.
