warning: not reporting region error due to nll
  --> $DIR/ex1-return-one-existing-name-return-type-is-anon.rs:18:5
   |
LL |     x //~ ERROR lifetime mismatch
   |     ^

error: unsatisfied lifetime constraints
  --> $DIR/ex1-return-one-existing-name-return-type-is-anon.rs:18:5
   |
LL |   fn foo<'a>(&self, x: &'a i32) -> &i32 {
   |          --  - let's call the lifetime of this reference `'1`
   |          |
   |          lifetime `'a` defined here
LL | 
LL |     x //~ ERROR lifetime mismatch
   |     ^ function was supposed to return data with lifetime `'a` but it is returning data with lifetime `'1`

error: aborting due to previous error

