warning: not reporting region error due to nll
  --> $DIR/ex1-return-one-existing-name-if-else-2.rs:12:16
   |
LL |     if x > y { x } else { y } //~ ERROR explicit lifetime
   |                ^

error[E0621]: explicit lifetime required in the type of `x`
  --> $DIR/ex1-return-one-existing-name-if-else-2.rs:12:8
   |
LL | fn foo<'a>(x: &i32, y: &'a i32) -> &'a i32 {
   |               ---- help: add explicit lifetime `'a` to the type of `x`: `&'a i32`
LL |     if x > y { x } else { y } //~ ERROR explicit lifetime
   |        ^^^^^ lifetime `'a` required

error: aborting due to previous error

For more information about this error, try `rustc --explain E0621`.
