warning: not reporting region error due to nll
  --> $DIR/ex3-both-anon-regions-using-impl-items.rs:15:16
   |
LL |         x.push(y); //~ ERROR lifetime mismatch
   |                ^

error: unsatisfied lifetime constraints
  --> $DIR/ex3-both-anon-regions-using-impl-items.rs:15:9
   |
LL |     fn foo(x: &mut Vec<&u8>, y: &u8) {
   |                        -        - let's call the lifetime of this reference `'1`
   |                        |
   |                        let's call the lifetime of this reference `'2`
LL |         x.push(y); //~ ERROR lifetime mismatch
   |         ^^^^^^^^^ argument requires that `'1` must outlive `'2`

error: aborting due to previous error

