warning: not reporting region error due to nll
  --> $DIR/ex3-both-anon-regions-return-type-is-anon.rs:17:5
   |
LL |     x //~ ERROR lifetime mismatch
   |     ^

error: unsatisfied lifetime constraints
  --> $DIR/ex3-both-anon-regions-return-type-is-anon.rs:17:5
   |
LL |   fn foo<'a>(&self, x: &i32) -> &i32 {
   |              -         - let's call the lifetime of this reference `'1`
   |              |
   |              let's call the lifetime of this reference `'2`
LL |     x //~ ERROR lifetime mismatch
   |     ^ function was supposed to return data with lifetime `'1` but it is returning data with lifetime `'2`

error: aborting due to previous error

