warning: not reporting region error due to nll
  --> $DIR/ex3-both-anon-regions-2.rs:12:10
   |
LL |     *v = x; //~ ERROR lifetime mismatch
   |          ^

error: unsatisfied lifetime constraints
  --> $DIR/ex3-both-anon-regions-2.rs:12:5
   |
LL | fn foo(&mut (ref mut v, w): &mut (&u8, &u8), x: &u8) {
   |                                   -             - let's call the lifetime of this reference `'1`
   |                                   |
   |                                   let's call the lifetime of this reference `'2`
LL |     *v = x; //~ ERROR lifetime mismatch
   |     ^^^^^^ requires that `'1` must outlive `'2`

error: aborting due to previous error

