warning: not reporting region error due to -Znll
  --> $DIR/propagate-fail-to-approximate-longer-wrong-bounds.rs:51:9
   |
51 |         demand_y(x, y, x.get())
   |         ^^^^^^^^^^^^^^^^^^^^^^^

error: free region `ReFree(DefId(0/1:18 ~ propagate_fail_to_approximate_longer_wrong_bounds[317d]::supply[0]::{{closure}}[0]), BrAnon(2))` does not outlive free region `ReFree(DefId(0/1:18 ~ propagate_fail_to_approximate_longer_wrong_bounds[317d]::supply[0]::{{closure}}[0]), BrAnon(4))`
  --> $DIR/propagate-fail-to-approximate-longer-wrong-bounds.rs:51:18
   |
51 |         demand_y(x, y, x.get())
   |                  ^

note: No external requirements
  --> $DIR/propagate-fail-to-approximate-longer-wrong-bounds.rs:49:47
   |
49 |       establish_relationships(&cell_a, &cell_b, |_outlives1, _outlives2, x, y| {
   |  _______________________________________________^
50 | |         // Only works if 'x: 'y:
51 | |         demand_y(x, y, x.get())
52 | |         //~^ WARN not reporting region error due to -Znll
53 | |         //~| ERROR does not outlive free region
54 | |     });
   | |_____^
   |
   = note: defining type: DefId(0/1:18 ~ propagate_fail_to_approximate_longer_wrong_bounds[317d]::supply[0]::{{closure}}[0]) with closure substs [
               i16,
               for<'r, 's, 't0, 't1, 't2, 't3> extern "rust-call" fn((&ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 'r)) std::cell::Cell<&ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 's)) &'_#1r u32>, &ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't0)) std::cell::Cell<&ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't1)) &'_#2r u32>, &ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't2)) std::cell::Cell<&ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 's)) u32>, &ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't3)) std::cell::Cell<&ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't1)) u32>))
           ]

note: No external requirements
  --> $DIR/propagate-fail-to-approximate-longer-wrong-bounds.rs:48:1
   |
48 | / fn supply<'a, 'b>(cell_a: Cell<&'a u32>, cell_b: Cell<&'b u32>) {
49 | |     establish_relationships(&cell_a, &cell_b, |_outlives1, _outlives2, x, y| {
50 | |         // Only works if 'x: 'y:
51 | |         demand_y(x, y, x.get())
...  |
54 | |     });
55 | | }
   | |_^
   |
   = note: defining type: DefId(0/0:6 ~ propagate_fail_to_approximate_longer_wrong_bounds[317d]::supply[0]) with substs []

error: aborting due to previous error

