warning: not reporting region error due to -Znll
  --> $DIR/propagate-approximated-shorter-to-static-wrong-bound.rs:51:9
   |
51 |         demand_y(x, y, x.get())
   |         ^^^^^^^^^^^^^^^^^^^^^^^

note: External requirements
  --> $DIR/propagate-approximated-shorter-to-static-wrong-bound.rs:48:47
   |
48 |       establish_relationships(&cell_a, &cell_b, |_outlives1, _outlives2, x, y| {
   |  _______________________________________________^
49 | |         //~^ ERROR does not outlive free region
50 | |         // Only works if 'x: 'y:
51 | |         demand_y(x, y, x.get())
52 | |         //~^ WARNING not reporting region error due to -Znll
53 | |     });
   | |_____^
   |
   = note: defining type: DefId(0/1:18 ~ propagate_approximated_shorter_to_static_wrong_bound[317d]::supply[0]::{{closure}}[0]) with closure substs [
               i16,
               for<'r, 's, 't0, 't1, 't2, 't3> extern "rust-call" fn((&ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 'r)) std::cell::Cell<&'_#1r &ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 's)) u32>, &ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't0)) std::cell::Cell<&'_#2r &ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't1)) u32>, &ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't2)) std::cell::Cell<&ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 's)) u32>, &ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't3)) std::cell::Cell<&ReLateBound(DebruijnIndex { depth: 1 }, BrNamed(crate0:DefIndex(0:0), 't1)) u32>))
           ]
   = note: number of external vids: 3
   = note: where '_#1r: '_#0r

error: free region `ReFree(DefId(0/0:6 ~ propagate_approximated_shorter_to_static_wrong_bound[317d]::supply[0]), BrNamed(crate0:DefIndex(1:16), 'a))` does not outlive free region `ReStatic`
  --> $DIR/propagate-approximated-shorter-to-static-wrong-bound.rs:48:47
   |
48 |       establish_relationships(&cell_a, &cell_b, |_outlives1, _outlives2, x, y| {
   |  _______________________________________________^
49 | |         //~^ ERROR does not outlive free region
50 | |         // Only works if 'x: 'y:
51 | |         demand_y(x, y, x.get())
52 | |         //~^ WARNING not reporting region error due to -Znll
53 | |     });
   | |_____^

note: No external requirements
  --> $DIR/propagate-approximated-shorter-to-static-wrong-bound.rs:47:1
   |
47 | / fn supply<'a, 'b>(cell_a: Cell<&'a u32>, cell_b: Cell<&'b u32>) {
48 | |     establish_relationships(&cell_a, &cell_b, |_outlives1, _outlives2, x, y| {
49 | |         //~^ ERROR does not outlive free region
50 | |         // Only works if 'x: 'y:
...  |
53 | |     });
54 | | }
   | |_^
   |
   = note: defining type: DefId(0/0:6 ~ propagate_approximated_shorter_to_static_wrong_bound[317d]::supply[0]) with substs []

error: aborting due to previous error

