error[E0430]: `self` import can only appear once in an import list
  --> $DIR/E0430.rs:1:16
   |
LL | use std::fmt::{self, self}; //~ ERROR E0430
   |                ^^^^  ---- another `self` import appears here
   |                |
   |                can only appear once in an import list

error[E0252]: the name `fmt` is defined multiple times
  --> $DIR/E0430.rs:1:22
   |
LL | use std::fmt::{self, self}; //~ ERROR E0430
   |                ------^^^^
   |                |   | |
   |                |   | `fmt` reimported here
   |                |   help: remove unnecessary import
   |                previous import of the module `fmt` here
   |
   = note: `fmt` must be defined only once in the type namespace of this module

error: aborting due to 2 previous errors

Some errors occurred: E0252, E0430.
For more information about an error, try `rustc --explain E0252`.
