error: expected one of `(`, `async`, `const`, `default`, `extern`, `fn`, `type`, or `unsafe`, found `}`
  --> $DIR/issue-41155.rs:13:1
   |
LL |     pub
   |        - expected one of 8 possible tokens here
LL | } //~ ERROR expected one of
   | ^ unexpected token

error[E0412]: cannot find type `S` in this scope
  --> $DIR/issue-41155.rs:11:6
   |
LL | impl S { //~ ERROR cannot find type
   |      ^ not found in this scope

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0412`.
