error: incorrect close delimiter: `}`
  --> $DIR/issue-10636-2.rs:19:1
   |
19 | } //~ ERROR: incorrect close delimiter
   | ^
   |
note: unclosed delimiter
  --> $DIR/issue-10636-2.rs:15:15
   |
15 |     option.map(|some| 42; //~ NOTE: unclosed delimiter
   |               ^

error: expected one of `,`, `.`, `?`, or an operator, found `;`
  --> $DIR/issue-10636-2.rs:15:25
   |
15 |     option.map(|some| 42; //~ NOTE: unclosed delimiter
   |                         ^ expected one of `,`, `.`, `?`, or an operator here

error: expected expression, found `)`
  --> $DIR/issue-10636-2.rs:19:1
   |
19 | } //~ ERROR: incorrect close delimiter
   | ^

error: main function not found

error: aborting due to previous error(s)

