error[E0252]: the name `foo` is defined multiple times
  --> $DIR/duplicate.rs:15:9
   |
LL |     use a::foo;
   |         ------ previous import of the value `foo` here
LL |     use a::foo; //~ ERROR the name `foo` is defined multiple times
   |     ----^^^^^^-
   |     |   |
   |     |   `foo` reimported here
   |     help: remove unnecessary import
   |
   = note: `foo` must be defined only once in the value namespace of this module

error[E0659]: `foo` is ambiguous (glob import vs glob import in the same module)
  --> $DIR/duplicate.rs:46:15
   |
LL |     use self::foo::bar; //~ ERROR `foo` is ambiguous
   |               ^^^ ambiguous name
   |
note: `foo` could refer to the module imported here
  --> $DIR/duplicate.rs:43:9
   |
LL |     use self::m1::*;
   |         ^^^^^^^^^^^
   = help: consider adding an explicit import of `foo` to disambiguate
note: `foo` could also refer to the module imported here
  --> $DIR/duplicate.rs:44:9
   |
LL |     use self::m2::*;
   |         ^^^^^^^^^^^
   = help: consider adding an explicit import of `foo` to disambiguate

error[E0659]: `foo` is ambiguous (glob import vs glob import in the same module)
  --> $DIR/duplicate.rs:35:8
   |
LL |     f::foo(); //~ ERROR `foo` is ambiguous
   |        ^^^ ambiguous name
   |
note: `foo` could refer to the function imported here
  --> $DIR/duplicate.rs:24:13
   |
LL |     pub use a::*;
   |             ^^^^
   = help: consider adding an explicit import of `foo` to disambiguate
note: `foo` could also refer to the function imported here
  --> $DIR/duplicate.rs:25:13
   |
LL |     pub use b::*;
   |             ^^^^
   = help: consider adding an explicit import of `foo` to disambiguate

error[E0659]: `foo` is ambiguous (glob import vs glob import in the same module)
  --> $DIR/duplicate.rs:49:9
   |
LL |         foo::bar(); //~ ERROR `foo` is ambiguous
   |         ^^^ ambiguous name
   |
note: `foo` could refer to the module imported here
  --> $DIR/duplicate.rs:43:9
   |
LL |     use self::m1::*;
   |         ^^^^^^^^^^^
   = help: consider adding an explicit import of `foo` to disambiguate
note: `foo` could also refer to the module imported here
  --> $DIR/duplicate.rs:44:9
   |
LL |     use self::m2::*;
   |         ^^^^^^^^^^^
   = help: consider adding an explicit import of `foo` to disambiguate

error: aborting due to 4 previous errors

Some errors occurred: E0252, E0659.
For more information about an error, try `rustc --explain E0252`.
