error[E0659]: `bar` is ambiguous
  --> $DIR/macro-paths.rs:25:5
   |
LL |     bar::m! { //~ ERROR ambiguous
   |     ^^^^^^
   |
note: `bar` could refer to the name defined here
  --> $DIR/macro-paths.rs:26:9
   |
LL |         mod bar { pub use two_macros::m; }
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `bar` could also refer to the name imported here
  --> $DIR/macro-paths.rs:24:9
   |
LL |     use foo::*;
   |         ^^^^^^
   = note: macro-expanded items do not shadow when used in a macro invocation path

error[E0659]: `baz` is ambiguous
  --> $DIR/macro-paths.rs:35:5
   |
LL |     baz::m! { //~ ERROR ambiguous
   |     ^^^^^^
   |
note: `baz` could refer to the name defined here
  --> $DIR/macro-paths.rs:36:9
   |
LL |         mod baz { pub use two_macros::m; }
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `baz` could also refer to the name defined here
  --> $DIR/macro-paths.rs:30:1
   |
LL | / pub mod baz {
LL | |     pub use two_macros::m;
LL | | }
   | |_^
   = note: macro-expanded items do not shadow when used in a macro invocation path

error[E0601]: main function not found

error: aborting due to 3 previous errors

You've got a few errors: E0601, E0659
If you want more information on an error, try using "rustc --explain E0601"
