error: any use of this value will cause an error
  --> $DIR/const_raw_ptr_ops.rs:6:1
   |
LL | const X: bool = unsafe { &1 as *const i32 == &2 as *const i32 }; //~ ERROR any use of this
   | ^^^^^^^^^^^^^^^^^^^^^^^^^------------------------------------^^^
   |                          |
   |                          "pointer arithmetic or comparison" needs an rfc before being allowed inside constants
   |
   = note: #[deny(const_err)] on by default

error: any use of this value will cause an error
  --> $DIR/const_raw_ptr_ops.rs:12:1
   |
LL | const Y2: usize = unsafe { &1 as *const i32 as usize + 1 }; //~ ERROR any use of this
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^-----------------------------^^^
   |                            |
   |                            "pointer arithmetic or comparison" needs an rfc before being allowed inside constants

error: any use of this value will cause an error
  --> $DIR/const_raw_ptr_ops.rs:16:1
   |
LL | const Z2: i32 = unsafe { *(42 as *const i32) }; //~ ERROR any use of this value will cause
   | ^^^^^^^^^^^^^^^^^^^^^^^^^-------------------^^^
   |                          |
   |                          a memory access tried to interpret some bytes as a pointer

error: any use of this value will cause an error
  --> $DIR/const_raw_ptr_ops.rs:17:1
   |
LL | const Z3: i32 = unsafe { *(44 as *const i32) }; //~ ERROR any use of this value will cause
   | ^^^^^^^^^^^^^^^^^^^^^^^^^-------------------^^^
   |                          |
   |                          a memory access tried to interpret some bytes as a pointer

error: aborting due to 4 previous errors

