error[E0019]: constant contains unimplemented expression type
  --> $DIR/issue-52475.rs:6:9
   |
LL | /         while n < 5 { //~ ERROR constant contains unimplemented expression type
LL | |             n = (n + 1) % 5; //~ ERROR evaluation of constant value failed
LL | |             x = &0; // Materialize a new AllocId
LL | |         }
   | |_________^

warning: Constant evaluating a complex constant, this might take some time
  --> $DIR/issue-52475.rs:2:18
   |
LL |       let _ = [(); {
   |  __________________^
LL | |         //~^ WARNING Constant evaluating a complex constant, this might take some time
LL | |         let mut x = &0;
LL | |         let mut n = 0;
...  |
LL | |         0
LL | |     }];
   | |_____^

error[E0080]: evaluation of constant value failed
  --> $DIR/issue-52475.rs:7:17
   |
LL |             n = (n + 1) % 5; //~ ERROR evaluation of constant value failed
   |                 ^^^^^^^^^^^ duplicate interpreter state observed here, const evaluation will never terminate

error: aborting due to 2 previous errors

Some errors occurred: E0019, E0080.
For more information about an error, try `rustc --explain E0019`.
