error: any use of this value will cause an error
  --> $DIR/union-const-eval-field.rs:37:5
   |
LL |     const FIELD3: Field3 = unsafe { UNION.field3 }; //~ ERROR any use of this value
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attempted to read undefined bytes
   |
   = note: #[deny(const_err)] on by default

error: any use of this value will cause an error
  --> $DIR/union-const-eval-field.rs:37:5
   |
LL |     const FIELD3: Field3 = unsafe { UNION.field3 }; //~ ERROR any use of this value
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attempted to read undefined bytes

error[E0080]: erroneous constant used
  --> $DIR/union-const-eval-field.rs:39:5
   |
LL |     FIELD3
   |     ^^^^^^ referenced constant has errors

error[E0080]: evaluation of constant expression failed
  --> $DIR/union-const-eval-field.rs:46:5
   |
LL |     FIELD3
   |     ------ referenced constant has errors
...
LL |     assert_eq!(read_field3(), unsafe { UNION.field3 });
   |     ^^^^^^^^^^^-------------^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                |
   |                inside call to `read_field3`
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0080`.
