error[E0026]: variant `X::Y` does not have a field named `number`
  --> $DIR/issue-41314.rs:17:16
   |
17 |         X::Y { number } => {} //~ ERROR does not have a field named `number`
   |                ^^^^^^ variant `X::Y` does not have field `number`

error[E0027]: pattern does not mention field `0`
  --> $DIR/issue-41314.rs:17:9
   |
17 |         X::Y { number } => {} //~ ERROR does not have a field named `number`
   |         ^^^^^^^^^^^^^^^ missing field `0`
   |
   = note: trying to match a tuple variant with a struct variant pattern

error: aborting due to 2 previous errors

