error[E0277]: the trait bound `{integer}: Scalar` is not satisfied
  --> $DIR/issue-22645.rs:25:5
   |
25 |   b + 3 //~ ERROR E0277
   |     ^ the trait `Scalar` is not implemented for `{integer}`
   |
   = help: the following implementations were found:
             <f64 as Scalar>
   = note: required because of the requirements on the impl of `std::ops::Add<{integer}>` for `Bob`

error[E0308]: mismatched types
  --> $DIR/issue-22645.rs:25:3
   |
23 | fn main() {
   |           - expected `()` because of this default return type
24 |   let b = Bob + 3.5;
25 |   b + 3 //~ ERROR E0277
   |   ^^^^^ expected (), found struct `Bob`
   |
   = note: expected type `()`
              found type `Bob`

error: aborting due to previous error(s)

