fn main() -> Result<(), ()> {
    a(b(c(d(e(
        //~^ ERROR cannot find function `a` in this scope
        //~| ERROR cannot find function `b` in this scope
        //~| ERROR cannot find function `c` in this scope
        //~| ERROR cannot find function `d` in this scope
        //~| ERROR cannot find function `e` in this scope
        z????????????????????????????????????????????????????????????????????????????????????????
        ?????????????????????????????????????????????????????????????????????????????????????????
        ??????????????????????????????????????????????????????????????????
        //~^^^ ERROR cannot find value `z` in this scope
    )))))
}
