error[E0560]: union `U` has no field named `principle`
  --> $DIR/union-suggest-field.rs:10:17
   |
LL |     let u = U { principle: 0 };
   |                 ^^^^^^^^^ help: a field with a similar name exists: `principal`

error[E0609]: no field `principial` on type `U`
  --> $DIR/union-suggest-field.rs:14:15
   |
LL |     let w = u.principial; //~ ERROR no field `principial` on type `U`
   |               ^^^^^^^^^^ help: a field with a similar name exists: `principal`

error[E0615]: attempted to take value of method `calculate` on type `U`
  --> $DIR/union-suggest-field.rs:18:15
   |
LL |     let y = u.calculate; //~ ERROR attempted to take value of method `calculate` on type `U`
   |               ^^^^^^^^^
   |
   = help: maybe a `()` to call it is missing?

error: aborting due to 3 previous errors

Some errors occurred: E0560, E0609, E0615.
For more information about an error, try `rustc --explain E0560`.
