warning: constant evaluation error: nonexistent struct field
  --> $DIR/union-const-eval.rs:21:21
   |
21 |         let b: [u8; C.b]; //~ ERROR constant evaluation error
   |                     ^^^
   |
   = note: #[warn(const_err)] on by default

error[E0080]: constant evaluation error
  --> $DIR/union-const-eval.rs:21:21
   |
21 |         let b: [u8; C.b]; //~ ERROR constant evaluation error
   |                     ^^^ nonexistent struct field

error: aborting due to previous error

