error: stability attributes may not be used outside of the standard library
  --> $DIR/stability-attribute-non-staged-force-unstable.rs:13:1
   |
LL | #[unstable] //~ ERROR: stability attributes may not be used
   | ^^^^^^^^^^^

error: stability attributes may not be used outside of the standard library
  --> $DIR/stability-attribute-non-staged-force-unstable.rs:14:1
   |
LL | #[stable] //~ ERROR: stability attributes may not be used
   | ^^^^^^^^^

error: stability attributes may not be used outside of the standard library
  --> $DIR/stability-attribute-non-staged-force-unstable.rs:15:1
   |
LL | #[rustc_deprecated] //~ ERROR: stability attributes may not be used
   | ^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

