error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
  --> $DIR/typeck-default-trait-impl-cross-crate-coherence.rs:23:1
   |
LL | impl DefaultedTrait for (A,) { } //~ ERROR E0117
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ impl doesn't use types inside crate
   |
   = note: the impl does not reference any types defined in this crate
   = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
  --> $DIR/typeck-default-trait-impl-cross-crate-coherence.rs:26:1
   |
LL | impl !DefaultedTrait for (B,) { } //~ ERROR E0117
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ impl doesn't use types inside crate
   |
   = note: the impl does not reference any types defined in this crate
   = note: define and implement a trait or new type instead

error[E0321]: cross-crate traits with a default impl, like `lib::DefaultedTrait`, can only be implemented for a struct/enum type defined in the current crate
  --> $DIR/typeck-default-trait-impl-cross-crate-coherence.rs:30:1
   |
LL | impl DefaultedTrait for Box<C> { } //~ ERROR E0321
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't implement cross-crate trait for type in another crate

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
  --> $DIR/typeck-default-trait-impl-cross-crate-coherence.rs:31:1
   |
LL | impl DefaultedTrait for lib::Something<C> { } //~ ERROR E0117
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ impl doesn't use types inside crate
   |
   = note: the impl does not reference any types defined in this crate
   = note: define and implement a trait or new type instead

error: aborting due to 4 previous errors

Some errors occurred: E0117, E0321.
For more information about an error, try `rustc --explain E0117`.
