error[E0308]: mismatched types
  --> $DIR/issue-26480.rs:26:19
   |
26 |                   $arr.len() * size_of($arr[0])); //~ ERROR mismatched types
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected u64, found usize
...
37 |     write!(hello);
   |     -------------- in this macro invocation

error[E0605]: non-primitive cast: `{integer}` as `()`
  --> $DIR/issue-26480.rs:32:19
   |
32 |     ($x:expr) => ($x as ()) //~ ERROR non-primitive cast
   |                   ^^^^^^^^
...
38 |     cast!(2);
   |     --------- in this macro invocation
   |
   = note: an `as` expression can only be used to convert between primitive types. Consider using the `From` trait

error: aborting due to 2 previous errors

