fn main() {
    a.5.2E+
    //~^ ERROR: unexpected token: `5.2E+`
    //~| ERROR: expected one of `.`, `;`, `?`, `}`, or an operator, found `5.2E+`
    //~| ERROR: expected at least one digit in exponent
}
