error[E0423]: expected value, found struct `Z`
  --> $DIR/privacy-struct-ctor.rs:26:9
   |
26 |         Z;
   |         ^
   |         |
   |         did you mean `Z { /* fields */ }`?
   |         did you mean `S`?
   |         constructor is not visible here due to private fields
   |
help: possible better candidate is found in another module, you can import it into scope
   |
15 | use m::n::Z;
   | ^^^^^^^^^^^^

error[E0423]: expected value, found struct `S`
  --> $DIR/privacy-struct-ctor.rs:36:5
   |
36 |     S;
   |     ^
   |     |
   |     did you mean `S { /* fields */ }`?
   |     constructor is not visible here due to private fields
   |
help: possible better candidate is found in another module, you can import it into scope
   |
13 | use m::S;
   | ^^^^^^^^^

error[E0423]: expected value, found struct `xcrate::S`
  --> $DIR/privacy-struct-ctor.rs:42:5
   |
42 |     xcrate::S;
   |     ^^^^^^^^^
   |     |
   |     did you mean `xcrate::S { /* fields */ }`?
   |     constructor is not visible here due to private fields
   |
help: possible better candidate is found in another module, you can import it into scope
   |
13 | use m::S;
   | ^^^^^^^^^

error[E0603]: tuple struct `Z` is private
  --> $DIR/privacy-struct-ctor.rs:25:9
   |
25 |         n::Z; //~ ERROR tuple struct `Z` is private
   |         ^^^^

error[E0603]: tuple struct `S` is private
  --> $DIR/privacy-struct-ctor.rs:35:5
   |
35 |     m::S; //~ ERROR tuple struct `S` is private
   |     ^^^^

error[E0603]: tuple struct `Z` is private
  --> $DIR/privacy-struct-ctor.rs:39:5
   |
39 |     m::n::Z; //~ ERROR tuple struct `Z` is private
   |     ^^^^^^^

error[E0603]: tuple struct `S` is private
  --> $DIR/privacy-struct-ctor.rs:41:5
   |
41 |     xcrate::m::S; //~ ERROR tuple struct `S` is private
   |     ^^^^^^^^^^^^

error[E0603]: tuple struct `Z` is private
  --> $DIR/privacy-struct-ctor.rs:45:5
   |
45 |     xcrate::m::n::Z; //~ ERROR tuple struct `Z` is private
   |     ^^^^^^^^^^^^^^^

error: aborting due to 8 previous errors

