error: incorrect close delimiter: `}`
  --> $DIR/token-error-correct-3.rs:31:9
   |
31 |         } else { //~ ERROR: incorrect close delimiter: `}`
   |         ^
   |
note: unclosed delimiter
  --> $DIR/token-error-correct-3.rs:23:21
   |
23 |             callback(path.as_ref();  //~ NOTE: unclosed delimiter
   |                     ^

error: expected one of `,`, `.`, `?`, or an operator, found `;`
  --> $DIR/token-error-correct-3.rs:23:35
   |
23 |             callback(path.as_ref();  //~ NOTE: unclosed delimiter
   |                                   ^ expected one of `,`, `.`, `?`, or an operator here

error: expected one of `.`, `;`, `?`, `}`, or an operator, found `)`
  --> $DIR/token-error-correct-3.rs:31:9
   |
26 |             fs::create_dir_all(path.as_ref()).map(|()| true) //~ ERROR: mismatched types
   |                                                             - expected one of `.`, `;`, `?`, `}`, or an operator here
...
31 |         } else { //~ ERROR: incorrect close delimiter: `}`
   |         ^ unexpected token

error[E0425]: cannot find function `is_directory` in this scope
  --> $DIR/token-error-correct-3.rs:21:13
   |
21 |         if !is_directory(path.as_ref()) { //~ ERROR: cannot find function `is_directory`
   |             ^^^^^^^^^^^^ not found in this scope

error[E0308]: mismatched types
  --> $DIR/token-error-correct-3.rs:26:13
   |
26 |             fs::create_dir_all(path.as_ref()).map(|()| true) //~ ERROR: mismatched types
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^- help: try adding a semicolon: `;`
   |             |
   |             expected (), found enum `std::result::Result`
   |
   = note: expected type `()`
              found type `std::result::Result<bool, std::io::Error>`

error: aborting due to 5 previous errors

