error[E0309]: the parameter type `T` may not live long enough
  --> $DIR/regions-enum-not-wf.rs:25:31
   |
LL | type RequireOutlives<'a, T> = <T as Dummy<'a>>::Out; //~ ERROR the parameter type `T` may not live long enough
   |                          -    ^^^^^^^^^^^^^^^^^^^^^
   |                          |
   |                          help: consider adding an explicit lifetime bound `T: 'a`...
   |
note: ...so that the type `T` will meet its required lifetime bounds
  --> $DIR/regions-enum-not-wf.rs:25:31
   |
LL | type RequireOutlives<'a, T> = <T as Dummy<'a>>::Out; //~ ERROR the parameter type `T` may not live long enough
   |                               ^^^^^^^^^^^^^^^^^^^^^

error[E0309]: the parameter type `T` may not live long enough
  --> $DIR/regions-enum-not-wf.rs:28:18
   |
LL | enum Ref1<'a, T> {
   |               - help: consider adding an explicit lifetime bound `T: 'a`...
LL |     Ref1Variant1(RequireOutlives<'a, T>) //~ ERROR the parameter type `T` may not live long enough
   |                  ^^^^^^^^^^^^^^^^^^^^^^
   |
note: ...so that the type `T` will meet its required lifetime bounds
  --> $DIR/regions-enum-not-wf.rs:28:18
   |
LL |     Ref1Variant1(RequireOutlives<'a, T>) //~ ERROR the parameter type `T` may not live long enough
   |                  ^^^^^^^^^^^^^^^^^^^^^^

error[E0309]: the parameter type `T` may not live long enough
  --> $DIR/regions-enum-not-wf.rs:33:25
   |
LL | enum Ref2<'a, T> {
   |               - help: consider adding an explicit lifetime bound `T: 'a`...
LL |     Ref2Variant1,
LL |     Ref2Variant2(isize, RequireOutlives<'a, T>), //~ ERROR the parameter type `T` may not live long enough
   |                         ^^^^^^^^^^^^^^^^^^^^^^
   |
note: ...so that the type `T` will meet its required lifetime bounds
  --> $DIR/regions-enum-not-wf.rs:33:25
   |
LL |     Ref2Variant2(isize, RequireOutlives<'a, T>), //~ ERROR the parameter type `T` may not live long enough
   |                         ^^^^^^^^^^^^^^^^^^^^^^

error[E0309]: the parameter type `T` may not live long enough
  --> $DIR/regions-enum-not-wf.rs:45:1
   |
LL |   enum RefDouble<'a, 'b, T> { //~ ERROR 45:1: 48:2: the parameter type `T` may not live long enough [E0309]
   |   ^                      - help: consider adding an explicit lifetime bound `T: 'b`...
   |  _|
   | |
LL | |     RefDoubleVariant1(&'a RequireOutlives<'b, T>)
LL | |         //~^ 46:23: 46:49: the parameter type `T` may not live long enough [E0309]
LL | | }
   | |_^
   |
note: ...so that the type `T` will meet its required lifetime bounds
  --> $DIR/regions-enum-not-wf.rs:45:1
   |
LL | / enum RefDouble<'a, 'b, T> { //~ ERROR 45:1: 48:2: the parameter type `T` may not live long enough [E0309]
LL | |     RefDoubleVariant1(&'a RequireOutlives<'b, T>)
LL | |         //~^ 46:23: 46:49: the parameter type `T` may not live long enough [E0309]
LL | | }
   | |_^

error[E0309]: the parameter type `T` may not live long enough
  --> $DIR/regions-enum-not-wf.rs:46:23
   |
LL | enum RefDouble<'a, 'b, T> { //~ ERROR 45:1: 48:2: the parameter type `T` may not live long enough [E0309]
   |                        - help: consider adding an explicit lifetime bound `T: 'b`...
LL |     RefDoubleVariant1(&'a RequireOutlives<'b, T>)
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: ...so that the type `T` will meet its required lifetime bounds
  --> $DIR/regions-enum-not-wf.rs:46:23
   |
LL |     RefDoubleVariant1(&'a RequireOutlives<'b, T>)
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0309`.
