error: lifetime name `'x` only used once
  --> $DIR/single_use_lifetimes-4.rs:12:12
   |
LL | struct Foo<'x> { //~ ERROR lifetime name `'x` only used once
   |            ^^
   |
note: lint level defined here
  --> $DIR/single_use_lifetimes-4.rs:10:9
   |
LL | #![deny(single_use_lifetimes)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: lifetime name `'x` only used once
  --> $DIR/single_use_lifetimes-4.rs:16:10
   |
LL | enum Bar<'x> { //~ ERROR lifetime name `'x` only used once
   |          ^^

error: aborting due to 2 previous errors

