error: lifetime name `'x` only used once
  --> $DIR/single_use_lifetimes-3.rs:11:12
   |
LL | struct Foo<'x> { //~ ERROR lifetime name `'x` only used once
   |            ^^
   |
note: lint level defined here
  --> $DIR/single_use_lifetimes-3.rs:10:9
   |
LL | #![deny(single_use_lifetimes)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: lifetime name `'y` only used once
  --> $DIR/single_use_lifetimes-3.rs:16:6
   |
LL | impl<'y> Foo<'y> { //~ ERROR lifetime name `'y` only used once
   |      ^^

error: aborting due to 2 previous errors

