error[E0133]: use of mutable static is unsafe and requires unsafe function or block
  --> $DIR/static-mut-requires-unsafe.rs:14:5
   |
LL |     a += 3;         //~ ERROR: requires unsafe
   |     ^^^^^^ use of mutable static
   |
   = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

error[E0133]: use of mutable static is unsafe and requires unsafe function or block
  --> $DIR/static-mut-requires-unsafe.rs:15:5
   |
LL |     a = 4;          //~ ERROR: requires unsafe
   |     ^^^^^ use of mutable static
   |
   = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

error[E0133]: use of mutable static is unsafe and requires unsafe function or block
  --> $DIR/static-mut-requires-unsafe.rs:16:14
   |
LL |     let _b = a;     //~ ERROR: requires unsafe
   |              ^ use of mutable static
   |
   = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0133`.
