warning: not reporting region error due to nll
  --> $DIR/expect-region-supply-region.rs:28:13
   |
LL |         f = Some(x); //~ ERROR borrowed data cannot be stored outside of its closure
   |             ^^^^^^^

warning: not reporting region error due to nll
  --> $DIR/expect-region-supply-region.rs:38:13
   |
LL |         f = Some(x); //~ ERROR borrowed data cannot be stored outside of its closure
   |             ^^^^^^^

warning: not reporting region error due to nll
  --> $DIR/expect-region-supply-region.rs:47:33
   |
LL |     closure_expecting_bound(|x: &'x u32| {
   |                                 ^^^^^^^

warning: not reporting region error due to nll
  --> $DIR/expect-region-supply-region.rs:52:13
   |
LL |         f = Some(x);
   |             ^^^^^^^

error: unsatisfied lifetime constraints
  --> $DIR/expect-region-supply-region.rs:28:9
   |
LL |     let mut f: Option<&u32> = None;
   |         ----- lifetime `'2` appears in the type of `f`
LL |     closure_expecting_bound(|x| {
   |                              - lifetime `'1` appears in this argument
LL |         f = Some(x); //~ ERROR borrowed data cannot be stored outside of its closure
   |         ^^^^^^^^^^^ assignment requires that `'1` must outlive `'2`

error: unsatisfied lifetime constraints
  --> $DIR/expect-region-supply-region.rs:38:9
   |
LL |     let mut f: Option<&u32> = None;
   |         ----- lifetime `'2` appears in the type of `f`
LL |     closure_expecting_bound(|x: &u32| {
   |                                 - let's call the lifetime of this reference `'1`
LL |         f = Some(x); //~ ERROR borrowed data cannot be stored outside of its closure
   |         ^^^^^^^^^^^ assignment requires that `'1` must outlive `'2`

error: unsatisfied lifetime constraints
  --> $DIR/expect-region-supply-region.rs:52:9
   |
LL |     let mut f: Option<&u32> = None;
   |         ----- lifetime `'2` appears in the type of `f`
...
LL |     closure_expecting_bound(|x: &'x u32| {
   |                                 - let's call the lifetime of this reference `'1`
...
LL |         f = Some(x);
   |         ^^^^^^^^^^^ assignment requires that `'1` must outlive `'2`

error: aborting due to 3 previous errors

