error[E0392]: parameter `'a` is never used
  --> $DIR/variance-regions-unused-indirect.rs:13:10
   |
LL | enum Foo<'a> { //~ ERROR parameter `'a` is never used
   |          ^^ unused type parameter
   |
   = help: consider removing `'a` or using a marker such as `std::marker::PhantomData`

error[E0392]: parameter `'a` is never used
  --> $DIR/variance-regions-unused-indirect.rs:17:10
   |
LL | enum Bar<'a> { //~ ERROR parameter `'a` is never used
   |          ^^ unused type parameter
   |
   = help: consider removing `'a` or using a marker such as `std::marker::PhantomData`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0392`.
