error[E0208]: [+, +]
  --> $DIR/variance-trait-bounds.rs:26:1
   |
LL | / struct TestStruct<U,T:Setter<U>> { //~ ERROR [+, +]
LL | |     t: T, u: U
LL | | }
   | |_^

error[E0208]: [*, +]
  --> $DIR/variance-trait-bounds.rs:31:1
   |
LL | / enum TestEnum<U,T:Setter<U>> { //~ ERROR [*, +]
LL | |     Foo(T)
LL | | }
   | |_^

error[E0208]: [*, +]
  --> $DIR/variance-trait-bounds.rs:36:1
   |
LL | / struct TestContraStruct<U,T:Setter<U>> { //~ ERROR [*, +]
LL | |     t: T
LL | | }
   | |_^

error[E0208]: [*, +]
  --> $DIR/variance-trait-bounds.rs:41:1
   |
LL | / struct TestBox<U,T:Getter<U>+Setter<U>> { //~ ERROR [*, +]
LL | |     t: T
LL | | }
   | |_^

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0208`.
