error[E0208]: [-, -, -]
  --> $DIR/variance-regions-direct.rs:19:1
   |
LL | / struct Test2<'a, 'b, 'c> { //~ ERROR [-, -, -]
LL | |     x: &'a isize,
LL | |     y: &'b [isize],
LL | |     c: &'c str
LL | | }
   | |_^

error[E0208]: [+, +, +]
  --> $DIR/variance-regions-direct.rs:28:1
   |
LL | / struct Test3<'a, 'b, 'c> { //~ ERROR [+, +, +]
LL | |     x: extern "Rust" fn(&'a isize),
LL | |     y: extern "Rust" fn(&'b [isize]),
LL | |     c: extern "Rust" fn(&'c str),
LL | | }
   | |_^

error[E0208]: [-, o]
  --> $DIR/variance-regions-direct.rs:37:1
   |
LL | / struct Test4<'a, 'b:'a> { //~ ERROR [-, o]
LL | |     x: &'a mut &'b isize,
LL | | }
   | |_^

error[E0208]: [+, o]
  --> $DIR/variance-regions-direct.rs:45:1
   |
LL | / struct Test5<'a, 'b:'a> { //~ ERROR [+, o]
LL | |     x: extern "Rust" fn(&'a mut &'b isize),
LL | | }
   | |_^

error[E0208]: [-, o]
  --> $DIR/variance-regions-direct.rs:55:1
   |
LL | / struct Test6<'a, 'b:'a> { //~ ERROR [-, o]
LL | |     x: &'a mut extern "Rust" fn(&'b isize),
LL | | }
   | |_^

error[E0208]: [*]
  --> $DIR/variance-regions-direct.rs:62:1
   |
LL | / struct Test7<'a> { //~ ERROR [*]
LL | |     x: isize
LL | | }
   | |_^

error[E0208]: [+, -, o]
  --> $DIR/variance-regions-direct.rs:69:1
   |
LL | / enum Test8<'a, 'b, 'c:'b> { //~ ERROR [+, -, o]
LL | |     Test8A(extern "Rust" fn(&'a isize)),
LL | |     Test8B(&'b [isize]),
LL | |     Test8C(&'b mut &'c str),
LL | | }
   | |_^

error: aborting due to 7 previous errors

For more information about this error, try `rustc --explain E0208`.
