warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:16:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:19:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:22:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:25:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:30:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:34:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:38:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:42:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:47:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:51:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:55:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

warning: not reporting region error due to nll
  --> $DIR/variance-btree-invariant-types.rs:59:5
   |
LL |     v //~ ERROR mismatched types
   |     ^

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:16:5
   |
LL | fn iter_cov_key<'a, 'new>(v: IterMut<'a, &'static (), ()>) -> IterMut<'a, &'new (), ()> {
   |                     ---- lifetime `'new` defined here
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:19:5
   |
LL | fn iter_cov_val<'a, 'new>(v: IterMut<'a, (), &'static ()>) -> IterMut<'a, (), &'new ()> {
   |                     ---- lifetime `'new` defined here
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:22:5
   |
LL | fn iter_contra_key<'a, 'new>(v: IterMut<'a, &'new (), ()>) -> IterMut<'a, &'static (), ()> {
   |                        ---- lifetime `'new` defined here
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:25:5
   |
LL | fn iter_contra_val<'a, 'new>(v: IterMut<'a, (), &'new ()>) -> IterMut<'a, (), &'static ()> {
   |                        ---- lifetime `'new` defined here
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:30:5
   |
LL | fn occ_cov_key<'a, 'new>(v: OccupiedEntry<'a, &'static (), ()>)
   |                    ---- lifetime `'new` defined here
LL |                          -> OccupiedEntry<'a, &'new (), ()> {
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:34:5
   |
LL | fn occ_cov_val<'a, 'new>(v: OccupiedEntry<'a, (), &'static ()>)
   |                    ---- lifetime `'new` defined here
LL |                          -> OccupiedEntry<'a, (), &'new ()> {
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:38:5
   |
LL | fn occ_contra_key<'a, 'new>(v: OccupiedEntry<'a, &'new (), ()>)
   |                       ---- lifetime `'new` defined here
LL |                             -> OccupiedEntry<'a, &'static (), ()> {
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:42:5
   |
LL | fn occ_contra_val<'a, 'new>(v: OccupiedEntry<'a, (), &'new ()>)
   |                       ---- lifetime `'new` defined here
LL |                             -> OccupiedEntry<'a, (), &'static ()> {
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:47:5
   |
LL | fn vac_cov_key<'a, 'new>(v: VacantEntry<'a, &'static (), ()>)
   |                    ---- lifetime `'new` defined here
LL |                          -> VacantEntry<'a, &'new (), ()> {
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:51:5
   |
LL | fn vac_cov_val<'a, 'new>(v: VacantEntry<'a, (), &'static ()>)
   |                    ---- lifetime `'new` defined here
LL |                          -> VacantEntry<'a, (), &'new ()> {
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:55:5
   |
LL | fn vac_contra_key<'a, 'new>(v: VacantEntry<'a, &'new (), ()>)
   |                       ---- lifetime `'new` defined here
LL |                             -> VacantEntry<'a, &'static (), ()> {
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: unsatisfied lifetime constraints
  --> $DIR/variance-btree-invariant-types.rs:59:5
   |
LL | fn vac_contra_val<'a, 'new>(v: VacantEntry<'a, (), &'new ()>)
   |                       ---- lifetime `'new` defined here
LL |                             -> VacantEntry<'a, (), &'static ()> {
LL |     v //~ ERROR mismatched types
   |     ^ return requires that `'new` must outlive `'static`

error: aborting due to 12 previous errors

