error[E0308]: mismatched types
  --> $DIR/float-literal-inference-restrictions.rs:12:18
   |
LL |     let x: f32 = 1; //~ ERROR mismatched types
   |                  ^ expected f32, found integral variable
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/float-literal-inference-restrictions.rs:13:18
   |
LL |     let y: f32 = 1f64; //~ ERROR mismatched types
   |                  ^^^^ expected f32, found f64

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
