warning: not reporting region error due to nll
  --> $DIR/regions-bounded-method-type-parameters-cross-crate.rs:30:7
   |
LL |     a.bigger_region(b) //~ ERROR 30:7: 30:20: lifetime mismatch [E0623]
   |       ^^^^^^^^^^^^^

error: unsatisfied lifetime constraints
  --> $DIR/regions-bounded-method-type-parameters-cross-crate.rs:30:5
   |
LL | fn call_bigger_region<'x, 'y>(a: Inv<'x>, b: Inv<'y>) {
   |                       --  -- lifetime `'y` defined here
   |                       |
   |                       lifetime `'x` defined here
LL |     // Here the value provided for 'y is 'y, and hence 'y:'x does not hold.
LL |     a.bigger_region(b) //~ ERROR 30:7: 30:20: lifetime mismatch [E0623]
   |     ^^^^^^^^^^^^^^^^^^ argument requires that `'y` must outlive `'x`

error: aborting due to previous error

