error[E0658]: procedural macros cannot expand to macro definitions (see issue #54727)
  --> $DIR/more-gates.rs:7:1
   |
LL | #[attr2mac1]
   | ^^^^^^^^^^^^
   |
   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable

error[E0658]: procedural macros cannot expand to macro definitions (see issue #54727)
  --> $DIR/more-gates.rs:10:1
   |
LL | #[attr2mac2]
   | ^^^^^^^^^^^^
   |
   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable

error[E0658]: procedural macros cannot expand to macro definitions (see issue #54727)
  --> $DIR/more-gates.rs:14:1
   |
LL | mac2mac1!(); //~ ERROR: cannot expand to macro definitions
   | ^^^^^^^^^^^^
   |
   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable

error[E0658]: procedural macros cannot expand to macro definitions (see issue #54727)
  --> $DIR/more-gates.rs:15:1
   |
LL | mac2mac2!(); //~ ERROR: cannot expand to macro definitions
   | ^^^^^^^^^^^^
   |
   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable

error[E0658]: procedural macros cannot expand to macro definitions (see issue #54727)
  --> $DIR/more-gates.rs:17:1
   |
LL | tricky!();
   | ^^^^^^^^^^
   |
   = help: add #![feature(proc_macro_hygiene)] to the crate attributes to enable

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0658`.
