From 124a11ce086952a5794d5cfbaa45175809497b81 Mon Sep 17 00:00:00 2001
From: None <none@example.com>
Date: Sat, 18 Nov 2023 10:50:36 -0500
Subject: [PATCH] [core] Disable portable-simd test

---
 library/core/tests/lib.rs | 2 --
 1 file changed, 2 deletions(-)

diff --git a/library/core/tests/lib.rs b/library/core/tests/lib.rs
index b71786c..cf484d5 100644
--- a/library/core/tests/lib.rs
+++ b/library/core/tests/lib.rs
@@ -87,7 +87,6 @@
 #![feature(numfmt)]
 #![feature(pattern)]
 #![feature(pointer_is_aligned_to)]
-#![feature(portable_simd)]
 #![feature(ptr_metadata)]
 #![feature(slice_from_ptr_range)]
 #![feature(slice_internals)]
@@ -155,7 +154,6 @@ mod pin;
 mod pin_macro;
 mod ptr;
 mod result;
-mod simd;
 mod slice;
 mod str;
 mod str_lossy;
-- 
2.45.2
