diff --git a/src/tools/compiletest/src/header.rs b/src/tools/compiletest/src/header.rs
index 887d27fd6dca4..2c2239f2b83d1 100644
--- a/src/tools/compiletest/src/header.rs
+++ b/src/tools/compiletest/src/header.rs
@@ -806,8 +806,8 @@ pub fn make_test_description<R: Read>(
     cfg: Option<&str>,
 ) -> test::TestDesc {
     let mut ignore = false;
     #[cfg(not(bootstrap))]
-    let ignore_message: Option<String> = None;
+    let ignore_message: Option<&str> = None;
     let mut should_fail = false;

     let rustc_has_profiler_support = env::var_os("RUSTC_PROFILER_SUPPORT").is_some();
