#![allow(unused_labels)]

// FIXME(#108019): outdated Unicode table
// fn foo() {
//     'ğŸ¥º loop {
//         break
//     }
// }

fn bar() {
    'ğŸ± loop {
    //~^ ERROR labeled expression must be followed by `:`
    //~| ERROR lifetimes or labels cannot contain emojis
        break
    }
}

fn qux() {
    'ağŸ± loop {
    //~^ ERROR labeled expression must be followed by `:`
    //~| ERROR lifetimes or labels cannot contain emojis
        break
    }
}

fn quux() {
    '1ğŸ± loop {
    //~^ ERROR labeled expression must be followed by `:`
    //~| ERROR lifetimes or labels cannot start with a number
        break
    }
}

fn x<'ğŸ±>() -> &'ğŸ± () {
    //~^ ERROR lifetimes or labels cannot contain emojis
    //~| ERROR lifetimes or labels cannot contain emojis
    &()
}

fn y() {
    'ağŸ±: loop {}
    //~^ ERROR lifetimes or labels cannot contain emojis
}

fn main() {}
