fn invalid_emoji_usages() {
    let arrowâ†”ï¸ = "basic emoji"; //~ ERROR: identifiers cannot contain emoji
    // FIXME
    let planetğŸª = "basic emoji"; //~ ERROR: unknown start of token
    // FIXME
    let wirelessğŸ›œ = "basic emoji"; //~ ERROR: unknown start of token
    // FIXME
    let key1ï¸âƒ£ = "keycap sequence"; //~ ERROR: unknown start of token
                                    //~^ WARN: identifier contains uncommon Unicode codepoints
    let flagğŸ‡ºğŸ‡³ = "flag sequence"; //~ ERROR: identifiers cannot contain emoji
    let walesğŸ´ = "tag sequence"; //~ ERROR: identifiers cannot contain emoji
    let foldedğŸ™ğŸ¿ = "modifier sequence"; //~ ERROR: identifiers cannot contain emoji
}

fn main() {
    invalid_emoji_usages();
}
