#[diagnostic::on_move]
//~^ERROR malformed `diagnostic::on_move` attribute input
struct Foo;

fn takes_foo(_: Foo) {}

fn main() {
    let foo = Foo;
    takes_foo(foo);
    let bar = foo;
    //~^ERROR use of moved value: `foo`
}
