error[E0596]: cannot borrow immutable borrowed content `*x` as mutable
  --> $DIR/borrowck-object-mutability.rs:19:5
   |
17 | fn borrowed_receiver(x: &Foo) {
   |                         ---- use `&mut Foo` here to make mutable
18 |     x.borrowed();
19 |     x.borrowed_mut(); //~ ERROR cannot borrow
   |     ^ cannot borrow as mutable

error[E0596]: cannot borrow immutable `Box` content `*x` as mutable
  --> $DIR/borrowck-object-mutability.rs:29:5
   |
27 | fn owned_receiver(x: Box<Foo>) {
   |                   - consider changing this to `mut x`
28 |     x.borrowed();
29 |     x.borrowed_mut(); //~ ERROR cannot borrow
   |     ^ cannot borrow as mutable

error: aborting due to previous error(s)

