warning: unused variable: `theOtherTwo`
  --> $DIR/issue-24690.rs:23:9
   |
23 |     let theOtherTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/issue-24690.rs:18:9
   |
18 | #![warn(unused)]
   |         ^^^^^^
   = note: #[warn(unused_variables)] implied by #[warn(unused)]
   = note: to avoid this warning, consider using `_theOtherTwo` instead

warning: variable `theTwo` should have a snake case name such as `the_two`
  --> $DIR/issue-24690.rs:22:9
   |
22 |     let theTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^
   |
   = note: #[warn(non_snake_case)] on by default

warning: variable `theOtherTwo` should have a snake case name such as `the_other_two`
  --> $DIR/issue-24690.rs:23:9
   |
23 |     let theOtherTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^^^^^^

error: compilation successful
  --> $DIR/issue-24690.rs:21:1
   |
21 | / fn main() { //~ ERROR compilation successful
22 | |     let theTwo = 2; //~ WARN should have a snake case name
23 | |     let theOtherTwo = 2; //~ WARN should have a snake case name
24 | |     //~^ WARN unused variable
25 | |     println!("{}", theTwo);
26 | | }
   | |_^

