error[E0106]: missing lifetime specifier
  --> $DIR/underscore-lifetime-binders.rs:12:17
   |
LL | struct Baz<'a>(&'_ &'a u8); //~ ERROR missing lifetime specifier
   |                 ^^ expected lifetime parameter

error[E0106]: missing lifetime specifier
  --> $DIR/underscore-lifetime-binders.rs:14:10
   |
LL | impl Foo<'_> { //~ ERROR missing lifetime specifier
   |          ^^ expected lifetime parameter

error[E0262]: invalid lifetime parameter name: `'_`
  --> $DIR/underscore-lifetime-binders.rs:18:8
   |
LL | fn foo<'_> //~ ERROR invalid lifetime parameter name: `'_`
   |        ^^ '_ is a reserved lifetime name

error[E0262]: invalid lifetime parameter name: `'_`
  --> $DIR/underscore-lifetime-binders.rs:24:21
   |
LL | fn meh() -> Box<for<'_> Meh<'_>> //~ ERROR invalid lifetime parameter name: `'_`
   |                     ^^ '_ is a reserved lifetime name

error[E0106]: missing lifetime specifier
  --> $DIR/underscore-lifetime-binders.rs:24:29
   |
LL | fn meh() -> Box<for<'_> Meh<'_>> //~ ERROR invalid lifetime parameter name: `'_`
   |                             ^^ expected lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
   = help: consider giving it a 'static lifetime

error[E0106]: missing lifetime specifier
  --> $DIR/underscore-lifetime-binders.rs:30:35
   |
LL | fn foo2(_: &'_ u8, y: &'_ u8) -> &'_ u8 { y } //~ ERROR missing lifetime specifier
   |                                   ^^ expected lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from `_` or `y`

error: aborting due to 6 previous errors

Some errors occurred: E0106, E0262.
For more information about an error, try `rustc --explain E0106`.
