error[E0425]: cannot find value `bar` in this scope
  --> $DIR/tab.rs:14:2
   |
14 |     bar; //~ ERROR cannot find value `bar`
   |     ^^^ not found in this scope

error[E0308]: mismatched types
  --> $DIR/tab.rs:18:2
   |
17 | fn foo() {
   |          - help: try adding a return type: `-> &'static str`
18 |     "bar            boo" //~ ERROR mismatched types
   |     ^^^^^^^^^^^^^^^^^^^^ expected (), found reference
   |
   = note: expected type `()`
              found type `&'static str`

error: aborting due to 2 previous errors

