error: invalid width `7` for integer literal
  --> $DIR/unicode_2.rs:14:25
   |
14 |     let _ = ("a̐éö̲", 0u7); //~ ERROR invalid width
   |                     ^^^
   |
   = help: valid widths are 8, 16, 32, 64 and 128

error: invalid width `42` for integer literal
  --> $DIR/unicode_2.rs:15:20
   |
15 |     let _ = ("아あ", 1i42); //~ ERROR invalid width
   |                      ^^^^
   |
   = help: valid widths are 8, 16, 32, 64 and 128

error[E0425]: cannot find value `a̐é` in this scope
  --> $DIR/unicode_2.rs:16:13
   |
16 |     let _ = a̐é; //~ ERROR cannot find
   |             ^^ not found in this scope

error: aborting due to 3 previous errors

