// Test for unsafe attributes generated by a proc-macro.

//@ proc-macro: unsafe-attributes-pm-in-2024.rs
//@ ignore-backends: gcc

unsafe_attributes_pm_in_2024::missing_unsafe!(); //~ ERROR unsafe attribute used without unsafe

#[unsafe_attributes_pm_in_2024::attr_missing_unsafe] //~ ERROR unsafe attribute used without unsafe
pub fn bar() {}

#[derive(unsafe_attributes_pm_in_2024::AttrMissingUnsafe)] //~ ERROR unsafe attribute used without unsafe
struct Baz;

unsafe_attributes_pm_in_2024::macro_rules_missing_unsafe!(); //~ ERROR unsafe attribute used without unsafe

make_fn!();

fn main() {}
