error[E0119]: conflicting implementations of trait `std::convert::AsRef<Q>` for type `std::boxed::Box<Q>`:
  --> $DIR/conflict-with-std.rs:17:1
   |
17 | impl AsRef<Q> for Box<Q> { //~ ERROR conflicting implementations
   | ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `alloc`:
           - impl<T> std::convert::AsRef<T> for std::boxed::Box<T>
             where T: ?Sized;

error[E0119]: conflicting implementations of trait `std::convert::From<S>` for type `S`:
  --> $DIR/conflict-with-std.rs:24:1
   |
24 | impl From<S> for S { //~ ERROR conflicting implementations
   | ^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T> std::convert::From<T> for T;

error[E0119]: conflicting implementations of trait `std::convert::TryFrom<X>` for type `X`:
  --> $DIR/conflict-with-std.rs:31:1
   |
31 | impl TryFrom<X> for X { //~ ERROR conflicting implementations
   | ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T, U> std::convert::TryFrom<U> for T
             where T: std::convert::From<U>;

error: aborting due to 3 previous errors

