warning: not reporting region error due to nll
  --> $DIR/regions-glb-free-free.rs:25:13
   |
LL |             Flag { //~ ERROR 25:13: 30:14: explicit lifetime required in the type of `s` [E0621]
   |             ^^^^

error[E0621]: explicit lifetime required in the type of `s`
  --> $DIR/regions-glb-free-free.rs:26:23
   |
LL |         pub fn set_desc(self, s: &str) -> Flag<'a> {
   |                                  ---- help: add explicit lifetime `'a` to the type of `s`: `&'a str`
LL |             Flag { //~ ERROR 25:13: 30:14: explicit lifetime required in the type of `s` [E0621]
LL |                 name: self.name,
   |                       ^^^^^^^^^ lifetime `'a` required

error: aborting due to previous error

For more information about this error, try `rustc --explain E0621`.
