error[E0261]: use of undeclared lifetime name `'a`
  --> $DIR/regions-in-structs.rs:20:9
   |
LL |     a: &'a isize, //~ ERROR use of undeclared lifetime name `'a`
   |         ^^ undeclared lifetime

error[E0261]: use of undeclared lifetime name `'a`
  --> $DIR/regions-in-structs.rs:21:9
   |
LL |     b: &'a isize, //~ ERROR use of undeclared lifetime name `'a`
   |         ^^ undeclared lifetime

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0261`.
