error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'a` due to conflicting requirements
  --> $DIR/regions-normalize-in-where-clause-list.rs:22:1
   |
LL | / fn bar<'a, 'b>() //~ ERROR cannot infer
LL | |     where <() as Project<'a, 'b>>::Item : Eq
LL | | {
LL | | }
   | |_^
   |
note: first, the lifetime cannot outlive the lifetime 'a as defined on the function body at 22:8...
  --> $DIR/regions-normalize-in-where-clause-list.rs:22:8
   |
LL | fn bar<'a, 'b>() //~ ERROR cannot infer
   |        ^^
note: ...but the lifetime must also be valid for the lifetime 'b as defined on the function body at 22:12...
  --> $DIR/regions-normalize-in-where-clause-list.rs:22:12
   |
LL | fn bar<'a, 'b>() //~ ERROR cannot infer
   |            ^^
   = note: ...so that the types are compatible:
           expected Project<'a, 'b>
              found Project<'_, '_>

error: aborting due to previous error

For more information about this error, try `rustc --explain E0495`.
