struct ABigðŸ‘©â€ðŸ‘©â€ðŸ‘§â€ðŸ‘§Family; //~ ERROR identifiers cannot contain emojis
struct ðŸ‘€; //~ ERROR identifiers cannot contain emojis
impl ðŸ‘€ {
    fn full_of_âœ¨() -> ðŸ‘€ { //~ ERROR identifiers cannot contain emojis
        ðŸ‘€
    }
}
fn i_like_to_ðŸ˜…_a_lot() -> ðŸ‘€ { //~ ERROR identifiers cannot contain emojis
    ðŸ‘€::full_ofâœ¨() //~ ERROR no function or associated item named `full_ofâœ¨` found for struct `ðŸ‘€`
    //~^ ERROR identifiers cannot contain emojis
}
fn main() {
    let _ = i_like_to_ðŸ˜„_a_lot(); //~ ERROR cannot find function `i_like_to_ðŸ˜„_a_lot` in this scope
    //~^ ERROR identifiers cannot contain emojis
}

