struct ABigðŸ‘©â€ðŸ‘©â€ðŸ‘§â€ðŸ‘§Family; //~ ERROR identifiers cannot contain emoji
struct ðŸ‘€; //~ ERROR identifiers cannot contain emoji
impl ðŸ‘€ {
    fn full_of_âœ¨() -> ðŸ‘€ { //~ ERROR identifiers cannot contain emoji
        ðŸ‘€
    }
}
fn i_like_to_ðŸ˜…_a_lot() -> ðŸ‘€ { //~ ERROR identifiers cannot contain emoji
    ðŸ‘€::full_ofâœ¨() //~ ERROR no function or associated item named `full_ofâœ¨` found for struct `ðŸ‘€`
    //~^ ERROR identifiers cannot contain emoji
}
fn main() {
    let _ = i_like_to_ðŸ˜„_a_lot() âž– 4; //~ ERROR cannot find function `i_like_to_ðŸ˜„_a_lot` in this scope
    //~^ ERROR identifiers cannot contain emoji
    //~| ERROR unknown start of token: \u{2796}
}
