error: expected identifier, found reserved keyword `macro`
  --> $DIR/macro-keyword.rs:1:4
   |
LL | fn macro() {  //~ ERROR expected identifier, found reserved keyword `macro`
   |    ^^^^^ expected identifier, found reserved keyword
help: you can escape reserved keywords to use them as identifiers
   |
LL | fn r#macro() {  //~ ERROR expected identifier, found reserved keyword `macro`
   |    ^^^^^^^

error: aborting due to previous error

