error: expected type, found `{`
  --> $DIR/recover-enum2.rs:8:18
   |
LL |             abc: {}, //~ ERROR: expected type, found `{`
   |                  ^

error: expected one of `!`, `(`, `)`, `+`, `,`, `::`, or `<`, found `{`
  --> $DIR/recover-enum2.rs:27:22
   |
LL |             Nope(i32 {}) //~ ERROR: found `{`
   |                      ^ expected one of 7 possible tokens here

error: aborting due to 2 previous errors

