error: expected type, found `3`
  --> $DIR/struct-literal-in-while.rs:13:12
   |
LL |         x: 3    //~ ERROR expected type, found `3`
   |            ^ expecting a type here because of type ascription

error: expected one of `.`, `;`, `?`, `}`, or an operator, found `{`
  --> $DIR/struct-literal-in-while.rs:14:12
   |
LL |     }.hi() { //~ ERROR expected one of `.`, `;`, `?`, `}`, or an operator, found `{`
   |            ^ expected one of `.`, `;`, `?`, `}`, or an operator here

error[E0423]: expected value, found struct `Foo`
  --> $DIR/struct-literal-in-while.rs:12:11
   |
LL |     while Foo { //~ ERROR expected value, found struct `Foo`
   |           ^^^ did you mean `(Foo { /* fields */ })`?

error[E0599]: no method named `hi` found for type `()` in the current scope
  --> $DIR/struct-literal-in-while.rs:14:7
   |
LL |     }.hi() { //~ ERROR expected one of `.`, `;`, `?`, `}`, or an operator, found `{`
   |       ^^

error: aborting due to 4 previous errors

Some errors occurred: E0423, E0599.
For more information about an error, try `rustc --explain E0423`.
