error: chained comparison operators require parentheses
  --> $DIR/issue-40396.rs:2:37
   |
LL |     println!("{:?}", (0..13).collect<Vec<i32>>()); //~ ERROR chained comparison
   |                                     ^^^^^^^^
   |
   = help: use `::<...>` instead of `<...>` if you meant to specify type arguments
   = help: or use `(...)` if you meant to specify fn arguments

error: chained comparison operators require parentheses
  --> $DIR/issue-40396.rs:6:25
   |
LL |     println!("{:?}", Vec<i32>::new()); //~ ERROR chained comparison
   |                         ^^^^^^^
   |
   = help: use `::<...>` instead of `<...>` if you meant to specify type arguments
   = help: or use `(...)` if you meant to specify fn arguments

error: chained comparison operators require parentheses
  --> $DIR/issue-40396.rs:10:37
   |
LL |     println!("{:?}", (0..13).collect<Vec<i32>()); //~ ERROR chained comparison
   |                                     ^^^^^^^^
   |
   = help: use `::<...>` instead of `<...>` if you meant to specify type arguments
   = help: or use `(...)` if you meant to specify fn arguments

error: chained comparison operators require parentheses
  --> $DIR/issue-40396.rs:10:41
   |
LL |     println!("{:?}", (0..13).collect<Vec<i32>()); //~ ERROR chained comparison
   |                                         ^^^^^^
   |
   = help: use `::<...>` instead of `<...>` if you meant to specify type arguments
   = help: or use `(...)` if you meant to specify fn arguments

error: aborting due to 4 previous errors

