error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
  --> $DIR/issue-31424.rs:17:9
   |
LL |         (&mut self).bar(); //~ ERROR cannot borrow
   |         ^^^^^^^^^^^
   |         |
   |         cannot borrow as mutable
   |         try removing `&mut` here

error[E0596]: cannot borrow `self` as mutable, as it is not declared as mutable
  --> $DIR/issue-31424.rs:23:9
   |
LL |         (&mut self).bar(); //~ ERROR cannot borrow
   |         ^^^^^^^^^^^
   |         |
   |         cannot borrow as mutable
   |         try removing `&mut` here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0596`.
