error[E0277]: the trait bound `i8: Foo<i32>` is not satisfied
  --> $DIR/issue-39802-show-5-trait-impls.rs:34:5
   |
LL |     Foo::<i32>::bar(&1i8); //~ ERROR is not satisfied
   |     ^^^^^^^^^^^^^^^ the trait `Foo<i32>` is not implemented for `i8`
   |
   = help: the following implementations were found:
             <i8 as Foo<bool>>
             <i8 as Foo<u16>>
             <i8 as Foo<u32>>
             <i8 as Foo<u64>>
             <i8 as Foo<u8>>
note: required by `Foo::bar`
  --> $DIR/issue-39802-show-5-trait-impls.rs:12:5
   |
LL |     fn bar(&self){}
   |     ^^^^^^^^^^^^^

error[E0277]: the trait bound `u8: Foo<i32>` is not satisfied
  --> $DIR/issue-39802-show-5-trait-impls.rs:35:5
   |
LL |     Foo::<i32>::bar(&1u8); //~ ERROR is not satisfied
   |     ^^^^^^^^^^^^^^^ the trait `Foo<i32>` is not implemented for `u8`
   |
   = help: the following implementations were found:
             <u8 as Foo<bool>>
             <u8 as Foo<u16>>
             <u8 as Foo<u32>>
             <u8 as Foo<u64>>
note: required by `Foo::bar`
  --> $DIR/issue-39802-show-5-trait-impls.rs:12:5
   |
LL |     fn bar(&self){}
   |     ^^^^^^^^^^^^^

error[E0277]: the trait bound `bool: Foo<i32>` is not satisfied
  --> $DIR/issue-39802-show-5-trait-impls.rs:36:5
   |
LL |     Foo::<i32>::bar(&true); //~ ERROR is not satisfied
   |     ^^^^^^^^^^^^^^^ the trait `Foo<i32>` is not implemented for `bool`
   |
   = help: the following implementations were found:
             <bool as Foo<u16>>
             <bool as Foo<u32>>
             <bool as Foo<u64>>
             <bool as Foo<u8>>
           and 2 others
note: required by `Foo::bar`
  --> $DIR/issue-39802-show-5-trait-impls.rs:12:5
   |
LL |     fn bar(&self){}
   |     ^^^^^^^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
