error[E0658]: using the `?` macro Kleene operator for "at most one" repetition is unstable (see issue #48075)
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:17:10
   |
LL |     ($(a)?) => {}
   |          ^
   |
   = help: add #![feature(macro_at_most_once_rep)] to the crate attributes to enable

error: expected `*` or `+`
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:17:10
   |
LL |     ($(a)?) => {}
   |          ^

error: expected `*` or `+`
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:23:11
   |
LL |     ($(a),?) => {} //~ERROR expected `*` or `+`
   |           ^
   |
   = note: `?` is not a macro repetition operator

error[E0658]: using the `?` macro Kleene operator for "at most one" repetition is unstable (see issue #48075)
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:27:10
   |
LL |     ($(a)?+) => {}
   |          ^
   |
   = help: add #![feature(macro_at_most_once_rep)] to the crate attributes to enable

error: expected `*` or `+`
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:27:10
   |
LL |     ($(a)?+) => {}
   |          ^

error[E0658]: using the `?` macro Kleene operator for "at most one" repetition is unstable (see issue #48075)
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:33:10
   |
LL |     ($(a)?*) => {}
   |          ^
   |
   = help: add #![feature(macro_at_most_once_rep)] to the crate attributes to enable

error: expected `*` or `+`
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:33:10
   |
LL |     ($(a)?*) => {}
   |          ^

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:41:11
   |
LL | macro_rules! foo {
   | ---------------- when calling this macro
...
LL |     foo!(a?); //~ ERROR no rules expected the token `?`
   |           ^ no rules expected this token in macro call

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:42:11
   |
LL | macro_rules! foo {
   | ---------------- when calling this macro
...
LL |     foo!(a?a); //~ ERROR no rules expected the token `?`
   |           ^ no rules expected this token in macro call

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018-feature-gate.rs:43:11
   |
LL | macro_rules! foo {
   | ---------------- when calling this macro
...
LL |     foo!(a?a?a); //~ ERROR no rules expected the token `?`
   |           ^ no rules expected this token in macro call

error: aborting due to 10 previous errors

For more information about this error, try `rustc --explain E0658`.
