error: no rules expected the token `bcd`
  --> $DIR/trace_faulty_macros.rs:17:26
   |
17 |         my_faulty_macro!(bcd); //~ ERROR no rules
   |                          ^^^
...
43 |     my_faulty_macro!();
   |     ------------------- in this macro invocation

note: trace_macro
  --> $DIR/trace_faulty_macros.rs:43:5
   |
43 |     my_faulty_macro!();
   |     ^^^^^^^^^^^^^^^^^^^
   |
   = note: expanding `my_faulty_macro! {  }`
   = note: to `my_faulty_macro ! ( bcd ) ;`
   = note: expanding `my_faulty_macro! { bcd }`

error: recursion limit reached while expanding the macro `my_recursive_macro`
  --> $DIR/trace_faulty_macros.rs:32:9
   |
32 |         my_recursive_macro!(); //~ ERROR recursion limit
   |         ^^^^^^^^^^^^^^^^^^^^^^
...
44 |     my_recursive_macro!();
   |     ---------------------- in this macro invocation
   |
   = help: consider adding a `#![recursion_limit="8"]` attribute to your crate

note: trace_macro
  --> $DIR/trace_faulty_macros.rs:44:5
   |
44 |     my_recursive_macro!();
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: expanding `my_recursive_macro! {  }`
   = note: to `my_recursive_macro ! (  ) ;`
   = note: expanding `my_recursive_macro! {  }`
   = note: to `my_recursive_macro ! (  ) ;`
   = note: expanding `my_recursive_macro! {  }`
   = note: to `my_recursive_macro ! (  ) ;`
   = note: expanding `my_recursive_macro! {  }`
   = note: to `my_recursive_macro ! (  ) ;`
   = note: expanding `my_recursive_macro! {  }`
   = note: to `my_recursive_macro ! (  ) ;`

