error[E0433]: failed to resolve. Maybe a missing `extern crate bad;`?
  --> $DIR/test.rs:60:12
   |
LL |     pub(in bad::path) mod m1 {} //~ ERROR failed to resolve. Maybe a missing `extern crate bad;`?
   |            ^^^ Maybe a missing `extern crate bad;`?

error: visibilities can only be restricted to ancestor modules
  --> $DIR/test.rs:61:12
   |
LL |     pub(in foo) mod m2 {} //~ ERROR visibilities can only be restricted to ancestor modules
   |            ^^^

error[E0364]: `f` is private, and cannot be re-exported
  --> $DIR/test.rs:31:24
   |
LL |         pub(super) use foo::bar::f as g; //~ ERROR cannot be re-exported
   |                        ^^^^^^^^^^^^^^^^
   |
note: consider marking `f` as `pub` in the imported module
  --> $DIR/test.rs:31:24
   |
LL |         pub(super) use foo::bar::f as g; //~ ERROR cannot be re-exported
   |                        ^^^^^^^^^^^^^^^^

error[E0603]: struct `Crate` is private
  --> $DIR/test.rs:48:9
   |
LL |     use pub_restricted::Crate; //~ ERROR private
   |         ^^^^^^^^^^^^^^^^^^^^^

error[E0603]: function `f` is private
  --> $DIR/test.rs:40:9
   |
LL |     use foo::bar::f; //~ ERROR private
   |         ^^^^^^^^^^^

error[E0616]: field `x` of struct `foo::bar::S` is private
  --> $DIR/test.rs:41:5
   |
LL |     S::default().x; //~ ERROR private
   |     ^^^^^^^^^^^^^^

error[E0624]: method `f` is private
  --> $DIR/test.rs:42:18
   |
LL |     S::default().f(); //~ ERROR private
   |                  ^

error[E0624]: method `g` is private
  --> $DIR/test.rs:43:5
   |
LL |     S::g(); //~ ERROR private
   |     ^^^^

error[E0616]: field `y` of struct `pub_restricted::Universe` is private
  --> $DIR/test.rs:52:13
   |
LL |     let _ = u.y; //~ ERROR private
   |             ^^^

error[E0616]: field `z` of struct `pub_restricted::Universe` is private
  --> $DIR/test.rs:53:13
   |
LL |     let _ = u.z; //~ ERROR private
   |             ^^^

error[E0624]: method `g` is private
  --> $DIR/test.rs:55:7
   |
LL |     u.g(); //~ ERROR private
   |       ^

error[E0624]: method `h` is private
  --> $DIR/test.rs:56:7
   |
LL |     u.h(); //~ ERROR private
   |       ^

error: aborting due to 12 previous errors

Some errors occurred: E0364, E0433, E0603, E0616, E0624.
For more information about an error, try `rustc --explain E0364`.
