error: lifetime parameters must be declared prior to type parameters
  --> $DIR/suggest-move-lifetimes.rs:1:13
   |
LL | struct A<T, 'a> { //~ ERROR lifetime parameters must be declared
   |             ^^
help: move the lifetime parameter prior to the first type parameter
   |
LL | struct A<'a, T> { //~ ERROR lifetime parameters must be declared
   |          ^^^ --

error: lifetime parameters must be declared prior to type parameters
  --> $DIR/suggest-move-lifetimes.rs:5:13
   |
LL | struct B<T, 'a, U> { //~ ERROR lifetime parameters must be declared
   |             ^^
help: move the lifetime parameter prior to the first type parameter
   |
LL | struct B<'a, T, U> { //~ ERROR lifetime parameters must be declared
   |          ^^^ --

error: lifetime parameters must be declared prior to type parameters
  --> $DIR/suggest-move-lifetimes.rs:10:16
   |
LL | struct C<T, U, 'a> { //~ ERROR lifetime parameters must be declared
   |                ^^
help: move the lifetime parameter prior to the first type parameter
   |
LL | struct C<'a, T, U> { //~ ERROR lifetime parameters must be declared
   |          ^^^    --

error: lifetime parameters must be declared prior to type parameters
  --> $DIR/suggest-move-lifetimes.rs:15:16
   |
LL | struct D<T, U, 'a, 'b, V, 'c> { //~ ERROR lifetime parameters must be declared
   |                ^^  ^^     ^^
help: move the lifetime parameter prior to the first type parameter
   |
LL | struct D<'a, 'b, 'c, T, U, V> { //~ ERROR lifetime parameters must be declared
   |          ^^^ ^^^ ^^^    -- --

error: aborting due to 4 previous errors

