error: expected type, found `10`
  --> $DIR/type-ascription-instead-of-initializer.rs:12:31
   |
LL |     let x: Vec::with_capacity(10, 20);  //~ ERROR expected type, found `10`
   |         --                    ^^
   |         ||
   |         |help: use `=` if you meant to assign
   |         while parsing the type for `x`

error[E0061]: this function takes 1 parameter but 2 parameters were supplied
  --> $DIR/type-ascription-instead-of-initializer.rs:12:12
   |
LL |     let x: Vec::with_capacity(10, 20);  //~ ERROR expected type, found `10`
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 1 parameter

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0061`.
