// run-rustfix
#![feature(non_ascii_idents)]

struct â„ğ“Šğ“ˆğ“‰;

fn main() {
    let Ã¼ = Rust;
    //~^ ERROR cannot find value `Rust` in this scope
    let _ = u;
    //~^ ERROR cannot find value `u` in this scope
}
