error[E0308]: mismatched types
  --> $DIR/loop-labeled-break-value.rs:13:29
   |
LL |         let _: i32 = loop { break }; //~ ERROR mismatched types
   |                             ^^^^^ expected (), found i32
   |
   = note: expected type `()`
              found type `i32`

error[E0308]: mismatched types
  --> $DIR/loop-labeled-break-value.rs:16:37
   |
LL |         let _: i32 = 'inner: loop { break 'inner }; //~ ERROR mismatched types
   |                                     ^^^^^^^^^^^^ expected (), found i32
   |
   = note: expected type `()`
              found type `i32`

error[E0308]: mismatched types
  --> $DIR/loop-labeled-break-value.rs:19:45
   |
LL |         let _: i32 = 'inner2: loop { loop { break 'inner2 } }; //~ ERROR mismatched types
   |                                             ^^^^^^^^^^^^^ expected (), found i32
   |
   = note: expected type `()`
              found type `i32`

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0308`.
